<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Roy Francis • 08-Dec-2019" />


<title>Downloads</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="assets/lab.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><b>NBIS • Workshop on RNA-Seq</b></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="home_schedule.html">
    <span class="fa fa-clipboard-list"></span>
     
    Schedule
  </a>
</li>
<li>
  <a href="home_lab.html">
    <span class="fa fa-flask"></span>
     
    Lab
  </a>
</li>
<li>
  <a href="home_precourse.html">
    <span class="fa fa-pencil-alt"></span>
     
    Precourse
  </a>
</li>
<li>
  <a href="home_info.html">
    <span class="fa fa-info"></span>
     
    Info
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Downloads</h1>
<h3 class="subtitle">Workshop on RNA-Seq</h3>
<h4 class="author"><b>Roy Francis</b> • 08-Dec-2019</h4>

</div>


<!-- rmd lab header -->
<p><br></p>
<div id="genomic-data" class="section level1">
<h1><span class="header-section-number">1</span> Genomic data</h1>
<p>Reference genomic data for your projects are available from Ensembl. This is usually the latest build of the genome, transcriptome etc as well as the annotations in GTF or GFF format. Most common organisms are available from <a href="https://www.ensembl.org/index.html">ensembl.org</a>. You can select the organism and then click on <strong>Download FASTA</strong>/<strong>Download GTF/GFF</strong> which takes you to the FTP site.</p>
<p>You can also go directly to their FTP site <a href="ftp://ftp.ensembl.org/pub/release-96" class="uri">ftp://ftp.ensembl.org/pub/release-96</a> where you can select the type of data you need, and then select the organism. For eg; <strong>homo_sapiens</strong>, under which you find <strong>cdna</strong>, <strong>cds</strong>, <strong>dna</strong>, <strong>dna_index</strong>, <strong>ncrna</strong> and <strong>pep</strong>. Under <strong>dna</strong>, the FASTA files are available as full genome or as separate chromosomes. Each of them are again available as regular (repeat content as normal bases), soft-masked (sm, repeat content in lowercase) or repeat-masked (rm, repeat content as Ns). Full genomes are also available as <strong>primary assembly</strong> or <strong>top-level</strong>. <strong>Primary assembly</strong> is what most people would need. The <strong>top-level</strong> is much larger in size and contains non-chromosomal contigs, patches, haplotypes etc. This is significantly larger in size compared to the primary assembly.</p>
<p><i class="fas fa-exclamation-circle"></i> Clades such as metazoa, protists, bacteria, fungi and plants are available through separate ensembl websites. These are listed on <a href="http://ensemblgenomes.org/" class="uri">http://ensemblgenomes.org/</a>.</p>
</div>
<div id="biomart" class="section level1">
<h1><span class="header-section-number">2</span> Biomart</h1>
<div id="genes" class="section level2">
<h2><span class="header-section-number">2.1</span> Genes</h2>
<p>In this section, we will download annotation data using R package <strong>biomaRt</strong>. Annotations refer to known features (verified experimentally or predicted) in the genome. Usually, our features of interest in RNA-Seq are genes, their IDs, position in the genome, gene biotype (protein coding, non-coding etc) etc. We will also use the <strong>dplyr</strong> package to pipe data through functions.</p>
<pre class="r"><code>library(biomaRt)
library(dplyr)

listMarts()</code></pre>
<pre><code>##                biomart               version
## 1 ENSEMBL_MART_ENSEMBL      Ensembl Genes 98
## 2   ENSEMBL_MART_MOUSE      Mouse strains 98
## 3     ENSEMBL_MART_SNP  Ensembl Variation 98
## 4 ENSEMBL_MART_FUNCGEN Ensembl Regulation 98</code></pre>
<p>We will use the code below to find the name of the Human ensembl genes dataset under ensembl mart.</p>
<pre class="r"><code>mart &lt;- useMart(&quot;ENSEMBL_MART_ENSEMBL&quot;)
ds &lt;- as.data.frame(listDatasets(mart=mart))

# find all rows in dataset &#39;ds&#39; where column &#39;description&#39; contains the string &#39;human&#39;
ds[grepl(&quot;human&quot;,tolower(ds$description)),]</code></pre>
<pre><code>##                  dataset              description    version
## 85 hsapiens_gene_ensembl Human genes (GRCh38.p13) GRCh38.p13</code></pre>
<p>Now that we know the name of the dataset, we can list all the columns (filters) in this dataset.</p>
<pre class="r"><code>mart &lt;- useMart(&quot;ENSEMBL_MART_ENSEMBL&quot;)
mart &lt;- useDataset(mart=mart,dataset=&quot;hsapiens_gene_ensembl&quot;)
la &lt;- listAttributes(mart=mart)
head(la)</code></pre>
<pre><code>##                            name                  description         page
## 1               ensembl_gene_id               Gene stable ID feature_page
## 2       ensembl_gene_id_version       Gene stable ID version feature_page
## 3         ensembl_transcript_id         Transcript stable ID feature_page
## 4 ensembl_transcript_id_version Transcript stable ID version feature_page
## 5            ensembl_peptide_id            Protein stable ID feature_page
## 6    ensembl_peptide_id_version    Protein stable ID version feature_page</code></pre>
<p>One can also search for attributes of interest.</p>
<pre class="r"><code>searchAttributes(mart=mart,pattern=&quot;entrez&quot;)</code></pre>
<pre><code>##                      name                   description         page
## 58  entrezgene_trans_name EntrezGene transcript name ID feature_page
## 78 entrezgene_description         NCBI gene description feature_page
## 79   entrezgene_accession           NCBI gene accession feature_page
## 80          entrezgene_id                  NCBI gene ID feature_page</code></pre>
<p>We create a vector of our columns of interest.</p>
<pre class="r"><code>myattributes &lt;- c(&quot;ensembl_gene_id&quot;,
                  &quot;entrezgene&quot;,
                  &quot;external_gene_name&quot;,
                  &quot;chromosome_name&quot;,
                  &quot;start_position&quot;,
                  &quot;end_position&quot;,
                  &quot;strand&quot;,
                  &quot;gene_biotype&quot;,
                  &quot;description&quot;)</code></pre>
<p>We then use this to download our data. Note that this can be a slow step.</p>
<pre class="r"><code>mart &lt;- useMart(&quot;ENSEMBL_MART_ENSEMBL&quot;)
mart &lt;- useDataset(mart=mart,dataset=&quot;hsapiens_gene_ensembl&quot;)
bdata &lt;- getBM(mart=mart,attributes=myattributes,uniqueRows=T)
head(bdata)</code></pre>
<pre><code>  ensembl_gene_id entrezgene external_gene_name chromosome_name start_position
1 ENSG00000210049         NA              MT-TF              MT            577
2 ENSG00000211459         NA            MT-RNR1              MT            648
3 ENSG00000210077         NA              MT-TV              MT           1602
4 ENSG00000210082         NA            MT-RNR2              MT           1671
5 ENSG00000209082         NA             MT-TL1              MT           3230
6 ENSG00000198888       4535             MT-ND1              MT           3307
  end_position strand   gene_biotype
1          647      1        Mt_tRNA
2         1601      1        Mt_rRNA
3         1670      1        Mt_tRNA
4         3229      1        Mt_rRNA
5         3304      1        Mt_tRNA
6         4262      1 protein_coding
                                                                                               description
1                              mitochondrially encoded tRNA-Phe (UUU/C) [Source:HGNC Symbol;Acc:HGNC:7481]
2                                       mitochondrially encoded 12S RNA [Source:HGNC Symbol;Acc:HGNC:7470]
3                                mitochondrially encoded tRNA-Val (GUN) [Source:HGNC Symbol;Acc:HGNC:7500]
4                                       mitochondrially encoded 16S RNA [Source:HGNC Symbol;Acc:HGNC:7471]
5                            mitochondrially encoded tRNA-Leu (UUA/G) 1 [Source:HGNC Symbol;Acc:HGNC:7490]
6 mitochondrially encoded NADH:ubiquinone oxidoreductase core subunit 1 [Source:HGNC Symbol;Acc:HGNC:7455]</code></pre>
<p>We find that there are several duplicates for all the IDs. This needs to be fixed when this information is to be used downstream.</p>
<pre class="r"><code>sum(duplicated(bdata$ensembl_gene_id))
sum(duplicated(bdata$entrezgene))
sum(duplicated(bdata$external_gene_name))</code></pre>
<pre><code>252
45751
6417</code></pre>
<pre class="r"><code># arrange table by chr name and start position
bdata &lt;- dplyr::arrange(bdata,chromosome_name,start_position)
write.table(bdata,&quot;./data/human_genes.txt&quot;,row.names=F,quote=F,col.names=T,sep=&quot;\t&quot;)</code></pre>
</div>
<div id="transcript" class="section level2">
<h2><span class="header-section-number">2.2</span> Transcript</h2>
<p>Here we download transcript to gene mappings. Notice that we can specify the <code>mart</code> and <code>dataset</code> in the <code>useMart()</code> function.</p>
<pre class="r"><code>mart &lt;- useMart(biomart=&quot;ensembl&quot;,dataset=&quot;hsapiens_gene_ensembl&quot;)
t2g &lt;- getBM(attributes=c(&quot;ensembl_transcript_id&quot;,&quot;ensembl_gene_id&quot;,&quot;external_gene_name&quot;),mart=mart)
write.table(t2g,&quot;./data/human_transcripts.txt&quot;,row.names=F,quote=F,col.names=T,sep=&quot;\t&quot;)</code></pre>
<p>The transcipt information file is saved to a file and will be used in the lab on <a href="lab_kallisto.html">Kallisto</a>.</p>
</div>
<div id="gene-ontology" class="section level2">
<h2><span class="header-section-number">2.3</span> Gene ontology</h2>
<p>Similarly, we can get entrez gene ID to GO ID relationships. List all the GO related filters:</p>
<pre class="r"><code>mart &lt;- biomaRt::useMart(biomart=&quot;ensembl&quot;,dataset=&quot;hsapiens_gene_ensembl&quot;)
lf &lt;- listFilters(mart=mart)

# find all rows in dataset &#39;lf&#39; where column &#39;name&#39; contains the string &#39;go&#39;
lf[grepl(&quot;go&quot;,tolower(lf$name)),]</code></pre>
<pre><code>                   name                        description
1               with_go                      With GO ID(s)
2       with_goslim_goa              With GOSlim GOA ID(s)
3                    go         GO ID(s) [e.g. GO:0000002]
4            goslim_goa GOSlim GOA ID(s) [e.g. GO:0000003]
5        go_parent_term              Parent term accession
6        go_parent_name                   Parent term name
7      go_evidence_code                   GO Evidence code
8   with_cdingo_homolog            Orthologous Dingo Genes
9 with_ggorilla_homolog          Orthologous Gorilla Genes</code></pre>
<pre class="r"><code>mart &lt;- biomaRt::useMart(biomart=&quot;ensembl&quot;,dataset=&quot;hsapiens_gene_ensembl&quot;)
bdata &lt;- getBM(mart=mart,attributes=c(&quot;entrezgene&quot;,&quot;go&quot;,&quot;go_evidence_code&quot;),uniqueRows=T)
write.table(bdata,&quot;./data/go.txt&quot;,row.names=F,quote=F,col.names=T,sep=&quot;\t&quot;)</code></pre>
</div>
<div id="id-conversion" class="section level2">
<h2><span class="header-section-number">2.4</span> ID conversion</h2>
<p>We can also take a quick look at converting IDs. It is often desirable to convert a certain gene identifier to another (ensembl gene ID, entrez gene ID, gene ID). Sometimes, it may be necessary to convert gene IDs of one organism to another. biomaRt has a convenient function for this called <code>getLDS()</code>.</p>
<p>Here is an example where we convert a few mouse ensembl IDs to Human Hugo gene IDs.</p>
<pre class="r"><code>mouse_genes &lt;- c(&quot;ENSMUSG00000035847&quot;,&quot;ENSMUSG00000000214&quot;)
mouse &lt;- useMart(&quot;ensembl&quot;,dataset=&quot;mmusculus_gene_ensembl&quot;)
human &lt;- useMart(&quot;ensembl&quot;,dataset=&quot;hsapiens_gene_ensembl&quot;)
human_genes &lt;- getLDS(attributes=c(&quot;ensembl_gene_id&quot;),filters=&quot;ensembl_gene_id&quot;,values=mouse_genes,mart=mouse, attributesL=c(&quot;hgnc_symbol&quot;),martL=human,valuesL=&quot;hgnc_symbol&quot;,uniqueRows=F)[,1]</code></pre>
<pre><code>      Gene.stable.ID HGNC.symbol
1 ENSMUSG00000000214          TH
2 ENSMUSG00000035847         IDS</code></pre>
</div>
</div>
<div id="session-info" class="section level1">
<h1><span class="header-section-number">3</span> Session info</h1>
<pre><code>## R version 3.6.0 (2017-01-27)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 16.04.6 LTS
## 
## Matrix products: default
## BLAS:   /home/travis/R-bin/lib/R/lib/libRblas.so
## LAPACK: /home/travis/R-bin/lib/R/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] pheatmap_1.0.12             edgeR_3.26.8               
##  [3] limma_3.40.6                DESeq2_1.24.0              
##  [5] SummarizedExperiment_1.14.1 DelayedArray_0.10.0        
##  [7] BiocParallel_1.18.1         matrixStats_0.55.0         
##  [9] Biobase_2.44.0              GenomicRanges_1.36.1       
## [11] GenomeInfoDb_1.20.0         IRanges_2.18.3             
## [13] S4Vectors_0.22.1            BiocGenerics_0.30.0        
## [15] biomaRt_2.40.5              ggplot2_3.2.1              
## [17] formattable_0.2.0.1         kableExtra_1.1.0           
## [19] dplyr_0.8.3                 lubridate_1.7.4            
## [21] leaflet_2.0.3               yaml_2.2.0                 
## [23] captioner_2.2.3             bookdown_0.16              
## [25] knitr_1.26                 
## 
## loaded via a namespace (and not attached):
##  [1] colorspace_1.4-1       ggsignif_0.6.0         ellipsis_0.3.0        
##  [4] htmlTable_1.13.3       XVector_0.24.0         base64enc_0.1-3       
##  [7] rstudioapi_0.10        ggpubr_0.2.4           farver_2.0.1          
## [10] bit64_0.9-7            AnnotationDbi_1.46.1   xml2_1.2.2            
## [13] splines_3.6.0          geneplotter_1.62.0     zeallot_0.1.0         
## [16] Formula_1.2-3          jsonlite_1.6           annotate_1.62.0       
## [19] cluster_2.0.8          shiny_1.4.0            readr_1.3.1           
## [22] compiler_3.6.0         httr_1.4.1             backports_1.1.5       
## [25] assertthat_0.2.1       Matrix_1.2-17          fastmap_1.0.1         
## [28] lazyeval_0.2.2         later_1.0.0            acepack_1.4.1         
## [31] htmltools_0.4.0        prettyunits_1.0.2      tools_3.6.0           
## [34] gtable_0.3.0           glue_1.3.1             GenomeInfoDbData_1.2.1
## [37] Rcpp_1.0.3             vctrs_0.2.0            nlme_3.1-139          
## [40] crosstalk_1.0.0        xfun_0.11              stringr_1.4.0         
## [43] rvest_0.3.5            mime_0.7               lifecycle_0.1.0       
## [46] XML_3.98-1.20          zlibbioc_1.30.0        scales_1.1.0          
## [49] hms_0.5.2              promises_1.1.0         RColorBrewer_1.1-2    
## [52] curl_4.3               memoise_1.1.0          gridExtra_2.3         
## [55] rpart_4.1-15           latticeExtra_0.6-28    stringi_1.4.3         
## [58] RSQLite_2.1.4          genefilter_1.66.0      checkmate_1.9.4       
## [61] rlang_0.4.2            pkgconfig_2.0.3        bitops_1.0-6          
## [64] evaluate_0.14          lattice_0.20-38        purrr_0.3.3           
## [67] labeling_0.3           htmlwidgets_1.5.1      cowplot_1.0.0         
## [70] bit_1.1-14             tidyselect_0.2.5       magrittr_1.5          
## [73] R6_2.4.1               Hmisc_4.3-0            DBI_1.0.0             
## [76] mgcv_1.8-28            pillar_1.4.2           foreign_0.8-71        
## [79] withr_2.1.2            survival_2.44-1.1      RCurl_1.95-4.12       
## [82] nnet_7.3-12            tibble_2.1.3           crayon_1.3.4          
## [85] rmarkdown_1.18         progress_1.2.2         locfit_1.5-9.1        
## [88] grid_3.6.0             data.table_1.12.6      blob_1.2.0            
## [91] digest_0.6.23          webshot_0.5.2          xtable_1.8-4          
## [94] tidyr_1.0.0            httpuv_1.5.2           munsell_0.5.0         
## [97] viridisLite_0.3.0</code></pre>
</div>


<div class="footer">
<div class="container" style="padding-top:5px;padding-bottom:5px;">
<p>
<span style="float:left; vertical-align:middle">
<a href="https://nbis.se/">NBIS</a> • <a href="https://www.scilifelab.se/">SciLifeLab</a>
</span>
<span style="float:right; vertical-align:middle">
<span class="footericon" style="padding-right:4px; padding-left:4px">
<a href="https://nbis.se/"><i class="fas fa-globe-americas"></i></a>
</span>
<span class="footericon" style="padding-right:4px; padding-left:4px">
<a href="https://twitter.com/NBISwe"><i class="fab fa-twitter"></i></a>
</span>
<span class="footericon" style="padding-left:4px">
<a href="https://www.linkedin.com/company/nbisweden/"><i class="fab fa-linkedin"></i></a>
</span>
</span>
</p>
</div>
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
