<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Roy Francis • 08-Dec-2019" />


<title>Intro to R</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="assets/lab.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><b>NBIS • Workshop on RNA-Seq</b></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="home_schedule.html">
    <span class="fa fa-clipboard-list"></span>
     
    Schedule
  </a>
</li>
<li>
  <a href="home_lab.html">
    <span class="fa fa-flask"></span>
     
    Lab
  </a>
</li>
<li>
  <a href="home_precourse.html">
    <span class="fa fa-pencil-alt"></span>
     
    Precourse
  </a>
</li>
<li>
  <a href="home_info.html">
    <span class="fa fa-info"></span>
     
    Info
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Intro to R</h1>
<h3 class="subtitle">Workshop on RNA-Seq</h3>
<h4 class="author"><b>Roy Francis</b> • 08-Dec-2019</h4>

</div>


<!-- rmd lab header -->
<p><br></p>
<p>R is a programming language for statistical computing, and data wrangling. It is open-source, widely used in data science, has a wide range of functions and algorithms for graphing and data analyses.</p>
<div id="assignment-operator" class="section level1">
<h1><span class="header-section-number">1</span> Assignment operator</h1>
<p>Variables are assigned usually using the <code>&lt;-</code> operator. The <code>=</code> operator also works in a similar way for most part.</p>
<pre class="r"><code>x &lt;- 4
x = 4
x</code></pre>
<pre><code>## [1] 4</code></pre>
</div>
<div id="arithmetic-operators" class="section level1">
<h1><span class="header-section-number">2</span> Arithmetic operators</h1>
<p>The commonly used arithmetic operators are shown below returning a number.</p>
<pre class="r"><code>x &lt;- 4
y &lt;- 2

# add
x + y

# subtract
x - y

# multiply
x * y

# divide
x / y

# modulus
x %% y

# power
x ^ y</code></pre>
<pre><code>## [1] 6
## [1] 2
## [1] 8
## [1] 2
## [1] 0
## [1] 16</code></pre>
</div>
<div id="logical-operators" class="section level1">
<h1><span class="header-section-number">3</span> Logical operators</h1>
<p>Logical operators return a logical <code>TRUE</code> or <code>FALSE</code>.</p>
<pre class="r"><code># equal to?
x == y

# not equal to?
x != y

# greater than?
x &gt; y

# less than?
x &lt; y

# greater than or equal to?
x &gt;= y

# less than or equal to?
x &lt;= y</code></pre>
<pre><code>## [1] FALSE
## [1] TRUE
## [1] TRUE
## [1] FALSE
## [1] TRUE
## [1] FALSE</code></pre>
</div>
<div id="data-types" class="section level1">
<h1><span class="header-section-number">4</span> Data types</h1>
<pre class="r"><code>class(1)
class(&quot;hello&quot;)
class(T)</code></pre>
<pre><code>## [1] &quot;numeric&quot;
## [1] &quot;character&quot;
## [1] &quot;logical&quot;</code></pre>
<pre class="r"><code>x &lt;- c(2,3,4,5,6)
y &lt;- c(&quot;a&quot;,&quot;c&quot;,&quot;d&quot;,&quot;e&quot;)
z &lt;- factor(c(&quot;a&quot;,&quot;c&quot;,&quot;d&quot;,&quot;e&quot;))
class(z)</code></pre>
<pre><code>## [1] &quot;factor&quot;</code></pre>
<pre class="r"><code>x &lt;- matrix(c(2,3,4,5,6,7),nrow=3,ncol=2)
class(x)
str(x)</code></pre>
<pre><code>## [1] &quot;matrix&quot;
##  num [1:3, 1:2] 2 3 4 5 6 7</code></pre>
<pre class="r"><code>dfr &lt;- data.frame(x = 1:3, y = c(&quot;a&quot;, &quot;b&quot;, &quot;c&quot;))
print(dfr)</code></pre>
<pre><code>##   x y
## 1 1 a
## 2 2 b
## 3 3 c</code></pre>
<pre class="r"><code>class(dfr)</code></pre>
<pre><code>## [1] &quot;data.frame&quot;</code></pre>
<pre class="r"><code>str(dfr)</code></pre>
<pre><code>## &#39;data.frame&#39;:    3 obs. of  2 variables:
##  $ x: int  1 2 3
##  $ y: Factor w/ 3 levels &quot;a&quot;,&quot;b&quot;,&quot;c&quot;: 1 2 3</code></pre>
</div>
<div id="accessors" class="section level1">
<h1><span class="header-section-number">5</span> Accessors</h1>
<p>Vectors positions can be accessed using <code>[]</code>. R follows 1-based indexing.</p>
<pre class="r"><code>x &lt;- c(2,3,4,5,6)
x
x[2]</code></pre>
<pre><code>## [1] 2 3 4 5 6
## [1] 3</code></pre>
<p>Dataframe or matrix positions can be accessed using <code>[]</code> specifying row and column like <code>[row,column]</code>.</p>
<pre class="r"><code>dfr &lt;- data.frame(x = 1:3, y = c(&quot;a&quot;, &quot;b&quot;, &quot;c&quot;))
dfr
dfr[1,]
dfr[,1]
dfr[2,2]</code></pre>
<pre><code>##   x y
## 1 1 a
## 2 2 b
## 3 3 c
##   x y
## 1 1 a
## [1] 1 2 3
## [1] b
## Levels: a b c</code></pre>
</div>
<div id="functions" class="section level1">
<h1><span class="header-section-number">6</span> Functions</h1>
<pre class="r"><code># generate 10 random numbers between 1 and 200
x &lt;- sample(x=1:200,10)
x

# length
length(x)

# sum
sum(x)

# mean
mean(x)

# median
median(x)

# min
min(x)

# log
log(x)

# exponent
exp(x)

# square-root
sqrt(x)

# round
round(x)

# sort
sort(x)</code></pre>
<pre><code>##  [1]  72  97 200  50 121 129 166  23 126 133
## [1] 10
## [1] 1117
## [1] 111.7
## [1] 123.5
## [1] 23
##  [1] 4.276666 4.574711 5.298317 3.912023 4.795791 4.859812 5.111988 3.135494
##  [9] 4.836282 4.890349
##  [1] 1.858672e+31 1.338335e+42 7.225974e+86 5.184706e+21 3.545131e+52
##  [6] 1.056789e+56 1.238466e+72 9.744803e+09 5.261441e+54 5.769871e+57
##  [1]  8.485281  9.848858 14.142136  7.071068 11.000000 11.357817 12.884099
##  [8]  4.795832 11.224972 11.532563
##  [1]  72  97 200  50 121 129 166  23 126 133
##  [1]  23  50  72  97 121 126 129 133 166 200</code></pre>
<p>Some useful string functions.</p>
<pre class="r"><code>a &lt;- &quot;sunny&quot;
b &lt;- &quot;day&quot;

# join
paste(a, b)

# find a pattern
grep(&quot;sun&quot;, a)

# number of characters
nchar(&quot;sunny&quot;)

# to uppercase
toupper(&quot;sunny&quot;)

# to lowercase
tolower(&quot;SUNNY&quot;)

# replace pattern
sub(&quot;sun&quot;, &quot;fun&quot;, &quot;sunny&quot;)

# substring
substr(&quot;sunny&quot;, start=1, stop=3)</code></pre>
<pre><code>## [1] &quot;sunny day&quot;
## [1] 1
## [1] 5
## [1] &quot;SUNNY&quot;
## [1] &quot;sunny&quot;
## [1] &quot;funny&quot;
## [1] &quot;sun&quot;</code></pre>
<p>Some general functions</p>
<pre class="r"><code>print(&quot;hello&quot;)
print(&quot;world&quot;)
cat(&quot;hello&quot;)
cat(&quot; world&quot;)
cat(&quot;\nhello\nworld&quot;)</code></pre>
<pre><code>## [1] &quot;hello&quot;
## [1] &quot;world&quot;
## hello world
## hello
## world</code></pre>
</div>
<div id="merging" class="section level1">
<h1><span class="header-section-number">7</span> Merging</h1>
<p>Two strings can be joined together using <code>paste()</code>.</p>
<pre class="r"><code>a &lt;- &quot;sunny&quot;
b &lt;- &quot;day&quot;

paste(a, b)
paste(a, b, sep=&quot;-&quot;)</code></pre>
<pre><code>## [1] &quot;sunny day&quot;
## [1] &quot;sunny-day&quot;</code></pre>
<p>The function <code>c()</code> is used to concatenate objects.</p>
<pre class="r"><code>a &lt;- &quot;sunny&quot;
b &lt;- &quot;day&quot;

c(a,b)</code></pre>
<pre><code>## [1] &quot;sunny&quot; &quot;day&quot;</code></pre>
<p>The function <code>cbind()</code> is used to join two dataframes column-wise.</p>
<pre class="r"><code>dfr1 &lt;- data.frame(x = 1:3, y = c(&quot;a&quot;, &quot;b&quot;, &quot;c&quot;))
dfr2 &lt;- data.frame(p = 4:6, q = c(&quot;d&quot;, &quot;e&quot;, &quot;f&quot;))
dfr1
dfr2

cbind(dfr1,dfr2)</code></pre>
<pre><code>##   x y
## 1 1 a
## 2 2 b
## 3 3 c
##   p q
## 1 4 d
## 2 5 e
## 3 6 f
##   x y p q
## 1 1 a 4 d
## 2 2 b 5 e
## 3 3 c 6 f</code></pre>
<p>Similarily, <code>rbind()</code> is used to join two dataframes row-wise.</p>
<pre class="r"><code>dfr1 &lt;- data.frame(x = 1:3, y = c(&quot;a&quot;, &quot;b&quot;, &quot;c&quot;))
dfr2 &lt;- data.frame(x = 4:6, y = c(&quot;d&quot;, &quot;e&quot;, &quot;f&quot;))
dfr1
dfr2

rbind(dfr1,dfr2)</code></pre>
<pre><code>##   x y
## 1 1 a
## 2 2 b
## 3 3 c
##   x y
## 1 4 d
## 2 5 e
## 3 6 f
##   x y
## 1 1 a
## 2 2 b
## 3 3 c
## 4 4 d
## 5 5 e
## 6 6 f</code></pre>
<p>Two dataframes can be merged based on a shared column using the <code>merge()</code> function.</p>
<pre class="r"><code>dfr1 &lt;- data.frame(x = 1:4, p = c(&quot;a&quot;, &quot;b&quot;, &quot;c&quot;,&quot;d&quot;))
dfr2 &lt;- data.frame(x = 3:6, q = c(&quot;l&quot;, &quot;m&quot;, &quot;n&quot;,&quot;o&quot;))
dfr1
dfr2

merge(dfr1,dfr2,by=&quot;x&quot;)
merge(dfr1,dfr2,by=&quot;x&quot;,all.x=T)
merge(dfr1,dfr2,by=&quot;x&quot;,all.y=T)
merge(dfr1,dfr2,by=&quot;x&quot;,all=T)</code></pre>
<pre><code>##   x p
## 1 1 a
## 2 2 b
## 3 3 c
## 4 4 d
##   x q
## 1 3 l
## 2 4 m
## 3 5 n
## 4 6 o
##   x p q
## 1 3 c l
## 2 4 d m
##   x p    q
## 1 1 a &lt;NA&gt;
## 2 2 b &lt;NA&gt;
## 3 3 c    l
## 4 4 d    m
##   x    p q
## 1 3    c l
## 2 4    d m
## 3 5 &lt;NA&gt; n
## 4 6 &lt;NA&gt; o
##   x    p    q
## 1 1    a &lt;NA&gt;
## 2 2    b &lt;NA&gt;
## 3 3    c    l
## 4 4    d    m
## 5 5 &lt;NA&gt;    n
## 6 6 &lt;NA&gt;    o</code></pre>
</div>
<div id="packages" class="section level1">
<h1><span class="header-section-number">8</span> Packages</h1>
<p>R packages extend the functionality of base R. R packages are stored in repositories of which the most commonly used is called <a href="https://cran.r-project.org/">CRAN</a> (The Comprehensive R Archive Network).</p>
<p>Packages are installed using the function <code>install.packages()</code>. Let’s install the graphics and plotting package <code>ggplot2</code> which will be useful in later sections.</p>
<pre class="r"><code>install.packages(&quot;ggplot2&quot;,dependencies=TRUE)</code></pre>
<p>Packages on BioConductor can be installed as follows:</p>
<pre class="r"><code>source(&quot;https://bioconductor.org/biocLite.R&quot;)
biocLite(&quot;biomaRt&quot;)</code></pre>
<p>Packages on GitHub can be installed using the function <code>install_github()</code> from package <code>devtools</code>.</p>
<p>Packages can also be installed from a local zipped file by providing a local path ans setting <code>type=&quot;source&quot;</code>.</p>
<pre class="r"><code>install.packages(&quot;./dir/package.zip&quot;,type=&quot;source&quot;)</code></pre>
<p>Inside RStudio, installing packages is much easier. Go to the <strong>Packages</strong> tab and click <strong>Install</strong>. In the window that opens up, you can find your package by typing into the <strong>Packages</strong> field and clicking <strong>Install</strong>. Bioconductor packages can be added to this list by setting it using <code>setRepositories()</code>.</p>
</div>
<div id="graphics" class="section level1">
<h1><span class="header-section-number">9</span> Graphics</h1>
<div id="base" class="section level2">
<h2><span class="header-section-number">9.1</span> Base</h2>
<p>R is an excellent tool for creating graphs and plots. The graphic capabilities and functions provided by the base R installation is called the base R graphics. Numerous packages exist to extend the functionality of base graphics.</p>
<p>We can try out plotting a few of the common plot types. Let’s start with a scatterplot. First we create a <code>data.frame</code> as this is the most commonly used data object.</p>
<pre class="r"><code>dfr &lt;- data.frame(a=sample(1:100,10),b=sample(1:100,10))</code></pre>
<p>Now we have a dataframe with two continuous variables that can be plotted against each other.</p>
<pre class="r"><code>plot(dfr$a,dfr$b)</code></pre>
<p><img src="lab_r_files/figure-html/unnamed-chunk-24-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
<p>This is probably the simplest and most basic plots. We can modify the x and y axis labels.</p>
<pre class="r"><code>plot(dfr$a,dfr$b,xlab=&quot;Variable a&quot;,ylab=&quot;Variable b&quot;)</code></pre>
<p><img src="lab_r_files/figure-html/unnamed-chunk-25-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
<p>We can change the point to a line.</p>
<pre class="r"><code>plot(dfr$a,dfr$b,xlab=&quot;Variable a&quot;,ylab=&quot;Variable b&quot;,type=&quot;b&quot;)</code></pre>
<p><img src="lab_r_files/figure-html/unnamed-chunk-26-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
<p>Let’s add a categorical column to our dataframe.</p>
<pre class="r"><code>dfr$cat &lt;- rep(c(&quot;C1&quot;,&quot;C2&quot;),each=5)</code></pre>
<p>And then colour the points by category.</p>
<pre class="r"><code># subset data
dfr_c1 &lt;- subset(dfr,dfr$cat == &quot;C1&quot;)
dfr_c2 &lt;- subset(dfr,dfr$cat == &quot;C2&quot;)

plot(dfr_c1$a,dfr_c1$b,xlab=&quot;Variable a&quot;,ylab=&quot;Variable b&quot;,col=&quot;red&quot;,pch=1)
points(dfr_c2$a,dfr_c2$b,col=&quot;blue&quot;,pch=2)

legend(x=&quot;topright&quot;,legend=c(&quot;C1&quot;,&quot;C2&quot;),
       col=c(&quot;red&quot;,&quot;blue&quot;),pch=c(1,2))</code></pre>
<p><img src="lab_r_files/figure-html/unnamed-chunk-28-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
<p>Let’s create a barplot.</p>
<pre class="r"><code>ldr &lt;- data.frame(a=letters[1:10],b=sample(1:50,10))
barplot(ldr$b,names.arg=ldr$a)</code></pre>
<p><img src="lab_r_files/figure-html/unnamed-chunk-29-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</div>
<div id="grid" class="section level2">
<h2><span class="header-section-number">9.2</span> Grid</h2>
<p>Grid graphics have a completely different underlying framework compared to base graphics. Generally, base graphics and grid graphics cannot be plotted together. The most popular grid-graphics based plotting library is <strong>ggplot2</strong>.</p>
<p>Let’s create the same plot are before using <strong>ggplot2</strong>. Make sure you have the package installed.</p>
<pre class="r"><code>library(ggplot2)

ggplot(dfr,aes(x=a,y=b,colour=cat))+
  geom_point()+
  labs(x=&quot;Variable a&quot;,y=&quot;Variable b&quot;)</code></pre>
<p><img src="lab_r_files/figure-html/unnamed-chunk-30-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
<p>It is generally easier and more consistent to create plots using the ggplot2 package compared to the base graphics.</p>
<p>Let’s create a barplot as well.</p>
<pre class="r"><code>ggplot(ldr,aes(x=a,y=b))+
  geom_bar(stat=&quot;identity&quot;)</code></pre>
<p><img src="lab_r_files/figure-html/unnamed-chunk-31-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</div>
</div>
<div id="inputoutput" class="section level1">
<h1><span class="header-section-number">10</span> Input/Output</h1>
<p>Input and output of data and images is an important aspect with data analysis.</p>
<div id="text" class="section level2">
<h2><span class="header-section-number">10.1</span> Text</h2>
<p>Data can come in a variety of formats which needs to be read into R and converted to an R data type.</p>
<p>Text files are the most commonly used input. Text files can be read in using the function <code>read.table</code>. We have a sample file to use: <strong>iris.txt</strong>.</p>
<pre class="r"><code>dfr &lt;- read.table(&quot;iris.txt&quot;,header=TRUE,stringsAsFactors=F)</code></pre>
<p>This reads in a tab-delimited text file with a header. The argument <code>sep='\t'</code> is set by default to specify that the delimiter is a tab. <code>stringsAsFactors=F</code> setting ensures that character columns are not automatically converted to factors.</p>
<p>It’s always a good idea to check the data after import.</p>
<pre class="r"><code>head(dfr)</code></pre>
<pre class="r"><code>str(dfr)</code></pre>
<p>Check <code>?read.table</code> for other wrapper functions to read in text files.</p>
<p>Let’s filter this data.frame and create a new dataset.</p>
<pre class="r"><code>dfr1 &lt;- dfr[dfr$Species == &quot;setosa&quot;,]</code></pre>
<p>And we can write this as a text file.</p>
<pre class="r"><code>write.table(dfr1,&quot;iris-setosa.txt&quot;,sep=&quot;\t&quot;,row.names=F,quote=F)</code></pre>
<p><code>sep=&quot;\t&quot;</code> sets the delimiter to tab. <code>row.names=F</code> denotes that rownames should not be written. <code>quote=F</code> specifies that doubles must not be placed around strings.</p>
</div>
<div id="images" class="section level2">
<h2><span class="header-section-number">10.2</span> Images</h2>
<p>Let’s take a look at saving plots.</p>
</div>
<div id="base-graphics" class="section level2">
<h2><span class="header-section-number">10.3</span> Base graphics</h2>
<p>The general idea for saving plots is open a graphics device, create the plot and then close the device. We will use <strong>png</strong> here. Check out <code>?png</code> for the arguments and other devices.</p>
<pre class="r"><code>dfr &lt;- data.frame(a=sample(1:100,10),b=sample(1:100,10))

png(filename=&quot;plot-base.png&quot;)
plot(dfr$a,dfr$b)
dev.off()</code></pre>
</div>
<div id="ggplot2" class="section level2">
<h2><span class="header-section-number">10.4</span> ggplot2</h2>
<p>The same idea can be applied to ggplot2, but in a slightly different way. First save the file to a variable, and then export the plot.</p>
<pre class="r"><code>p &lt;- ggplot(dfr,aes(a,b)) + geom_point()

png(filename=&quot;plot-ggplot-1.png&quot;)
print(p)
dev.off()</code></pre>
<p><strong>ggplot2</strong> also has another easier helper function to export images.</p>
<pre class="r"><code>ggsave(filename=&quot;plot-ggplot-2.png&quot;,plot=p)</code></pre>
</div>
</div>
<div id="getting-help" class="section level1">
<h1><span class="header-section-number">11</span> Getting help</h1>
<ul>
<li>Use <code>?function</code> to get function documentation</li>
<li>Use <code>??bla</code> to search for a function</li>
<li>Use <code>args(function)</code> to get the arguments to a function</li>
<li>Go to the package CRAN page/webpage for vignettes</li>
<li><a href="http://www.cookbook-r.com/">R Cookbook</a>: General purpose reference.</li>
<li><a href="https://www.statmethods.net/">Quick R</a>: General purpose reference.</li>
<li><a href="https://stackoverflow.com/">Stackoverflow</a>: Online community to find solutions to your problems.</li>
</ul>
</div>
<div id="session-info" class="section level1">
<h1><span class="header-section-number">12</span> Session info</h1>
<pre><code>## R version 3.6.0 (2017-01-27)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 16.04.6 LTS
## 
## Matrix products: default
## BLAS:   /home/travis/R-bin/lib/R/lib/libRblas.so
## LAPACK: /home/travis/R-bin/lib/R/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] fgsea_1.10.1                Rcpp_1.0.3                 
##  [3] enrichR_2.1                 pvclust_2.2-0              
##  [5] rafalib_1.0.0               pheatmap_1.0.12            
##  [7] edgeR_3.26.8                limma_3.40.6               
##  [9] DESeq2_1.24.0               SummarizedExperiment_1.14.1
## [11] DelayedArray_0.10.0         BiocParallel_1.18.1        
## [13] matrixStats_0.55.0          Biobase_2.44.0             
## [15] GenomicRanges_1.36.1        GenomeInfoDb_1.20.0        
## [17] IRanges_2.18.3              S4Vectors_0.22.1           
## [19] BiocGenerics_0.30.0         biomaRt_2.40.5             
## [21] ggplot2_3.2.1               formattable_0.2.0.1        
## [23] kableExtra_1.1.0            dplyr_0.8.3                
## [25] lubridate_1.7.4             leaflet_2.0.3              
## [27] yaml_2.2.0                  captioner_2.2.3            
## [29] bookdown_0.16               knitr_1.26                 
## 
## loaded via a namespace (and not attached):
##  [1] colorspace_1.4-1       rjson_0.2.20           ggsignif_0.6.0        
##  [4] ellipsis_0.3.0         htmlTable_1.13.3       XVector_0.24.0        
##  [7] base64enc_0.1-3        rstudioapi_0.10        ggpubr_0.2.4          
## [10] farver_2.0.1           bit64_0.9-7            AnnotationDbi_1.46.1  
## [13] xml2_1.2.2             splines_3.6.0          geneplotter_1.62.0    
## [16] zeallot_0.1.0          Formula_1.2-3          jsonlite_1.6          
## [19] annotate_1.62.0        cluster_2.0.8          shiny_1.4.0           
## [22] readr_1.3.1            compiler_3.6.0         httr_1.4.1            
## [25] backports_1.1.5        assertthat_0.2.1       Matrix_1.2-17         
## [28] fastmap_1.0.1          lazyeval_0.2.2         later_1.0.0           
## [31] acepack_1.4.1          htmltools_0.4.0        prettyunits_1.0.2     
## [34] tools_3.6.0            gtable_0.3.0           glue_1.3.1            
## [37] GenomeInfoDbData_1.2.1 fastmatch_1.1-0        vctrs_0.2.0           
## [40] nlme_3.1-139           crosstalk_1.0.0        xfun_0.11             
## [43] stringr_1.4.0          rvest_0.3.5            mime_0.7              
## [46] lifecycle_0.1.0        XML_3.98-1.20          zlibbioc_1.30.0       
## [49] scales_1.1.0           hms_0.5.2              promises_1.1.0        
## [52] RColorBrewer_1.1-2     curl_4.3               memoise_1.1.0         
## [55] gridExtra_2.3          rpart_4.1-15           latticeExtra_0.6-28   
## [58] stringi_1.4.3          RSQLite_2.1.4          genefilter_1.66.0     
## [61] checkmate_1.9.4        rlang_0.4.2            pkgconfig_2.0.3       
## [64] bitops_1.0-6           evaluate_0.14          lattice_0.20-38       
## [67] purrr_0.3.3            labeling_0.3           htmlwidgets_1.5.1     
## [70] cowplot_1.0.0          bit_1.1-14             tidyselect_0.2.5      
## [73] magrittr_1.5           R6_2.4.1               Hmisc_4.3-0           
## [76] DBI_1.0.0              mgcv_1.8-28            pillar_1.4.2          
## [79] foreign_0.8-71         withr_2.1.2            survival_2.44-1.1     
## [82] RCurl_1.95-4.12        nnet_7.3-12            tibble_2.1.3          
## [85] crayon_1.3.4           rmarkdown_1.18         progress_1.2.2        
## [88] locfit_1.5-9.1         grid_3.6.0             data.table_1.12.6     
## [91] blob_1.2.0             digest_0.6.23          webshot_0.5.2         
## [94] xtable_1.8-4           tidyr_1.0.0            httpuv_1.5.2          
## [97] munsell_0.5.0          viridisLite_0.3.0</code></pre>
<p><strong>End of document</strong></p>
</div>


<div class="footer">
<div class="container" style="padding-top:5px;padding-bottom:5px;">
<p>
<span style="float:left; vertical-align:middle">
<a href="https://nbis.se/">NBIS</a> • <a href="https://www.scilifelab.se/">SciLifeLab</a>
</span>
<span style="float:right; vertical-align:middle">
<span class="footericon" style="padding-right:4px; padding-left:4px">
<a href="https://nbis.se/"><i class="fas fa-globe-americas"></i></a>
</span>
<span class="footericon" style="padding-right:4px; padding-left:4px">
<a href="https://twitter.com/NBISwe"><i class="fab fa-twitter"></i></a>
</span>
<span class="footericon" style="padding-left:4px">
<a href="https://www.linkedin.com/company/nbisweden/"><i class="fab fa-linkedin"></i></a>
</span>
</span>
</p>
</div>
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
