---
title: "<i class='fas fa-flask' style='margin-left:4px;margin-right:4px;'></i> Lab"
output:
  bookdown::html_document2:
    toc: true
    toc_float: true
    toc_depth: 4
    number_sections: true
    theme: flatly
    highlight: tango
    df_print: default
    code_folding: "none"
    self_contained: false
    keep_md: false
    encoding: 'UTF-8'
    css: "assets/lab.css"
---

```{r,child="assets/header-lab.Rmd"}
```

```{r,include=FALSE}
id_project <- "g2019011"
id_reservation_2 <- "g2019011_2"
id_reservation_3 <- "g2019011_3"
```

This page contains links to different tutorials that are used in this course. The tutorials are well documented and should be easy to follow.

<div class="instruction">
Input code blocks are displayed like shown below. The code language is displayed above the block. Shell scripts (**SH**) are to be executed in the linux terminal such as bash. **R** scripts are to be run in R either through the terminal, RGui or RStudio.

```{sh,eval=FALSE,block.title=TRUE}
command
```

<i class="fas fa-exclamation-circle"></i> Note &nbsp; <i class="fas fa-lightbulb"></i> Tip &nbsp; <i class="fas fa-comments"></i> Discuss &nbsp; <i class="fas fa-clipboard-list"></i> Task
</div>

<br>

# Introduction

Basic introduction to working in the Linux environment using a command-line interface.

[<i class="fas fa-link"></i> <span class="large">Introduction to Linux</span>](linux-intro.html)  

Handling and analysing large data may not be feasible on a local computer. We have access to a computing cluster called UPPMAX. Below are exercises to start working on UPPMAX.

[<i class="fas fa-link"></i> <span class="large">Introduction to Uppmax</span>](uppmax-intro.html)  

Most of the analyses is carried out in R and it will be useful to learn some basic R.

[<i class="fas fa-link"></i> <span class="large">Introduction to R</span>](lab_r.html)  

This topic covers retrieving supporting data needed for RNA-seq analyses. These include gene annotation IDs such as mapping between Ensembl IDs and Gene IDs, GO terms and transcript IDs. We also cover retrieving genomic data from Ensembl.

[<i class="fas fa-link"></i> <span class="large">Downloading data</span>](lab_download.html)

# Main lab
## Data

In most of the exercises, we will use RNA-seq data (Illumina short reads) from the human A431 cell line. It is an epidermoid carcinoma cell line which is often used to study cancer and the cell cycle, and as a sort of positive control of epidermal growth factor receptor (EGFR) expression. A431 cells express very high levels of EGFR, in contrast to normal human fibroblasts.

The A431 cells were treated with gefinitib, which is an EGFR inhibitor and is used (under the trade name Iressa) as a drug to treat cancers with mutated and overactive EGFR. In the experiment, RNA was extracted at four time points: before the gefinitib treatment (t=0), and two, six and twenty-four hours after treatment (t=2, t=6, t=24, respectively), and sequenced using an Illumina HiSeq instrument in triplicates (thus there are 3x4=12 samples).

This data set or parts of it will be used in the labs on read mapping, transcript assembly, visualization, quality control and differential expression. There are many relevant questions that could be asked based on these measurements. In the QC exercise, we are going to examine if the RNA libraries that we work with are what we think they are or if there is any mislabelling. In the isoform exercise, we are going to look at some specific regions where the mass-spectrometry data indicated that novel exons or splice variants could be present at the protein level. We will use (part of) the RNA-seq data to examine if there is corresponding evidence on the mRNA level, and how different software tools could be used to detect novel gene variants.

## Working on Uppmax

The step **Quality control**, **Mapping** and **Quantification** will be run on computing cluster **Rackham** which is part of UPPMAX. A standard compute node on cluster Rackham has 128 GB of RAM and 20 cores. Therefore, each core gives you 6.4 GB of RAM. We will use 8 cores per person for this session which gives you about 51 GB RAM. 

### Connecting

<i class="fas fa-lightbulb"></i> Log in to Uppmax in a way so that the generated graphics are exported via the network to your screen. Login in to Uppmax with X-forwarding enabled. This will allow any graphical interface that you start on your compute node to be exported to your computer.

Linux and Mac users will run this on the terminal. Windows users will run this in a tool such as MobaXterm.

```{sh,block.title=TRUE,eval=FALSE}
ssh -Y username@rackham.uppmax.uu.se
```

<i class="fas fa-exclamation-circle"></i> Note that Mac users will need to have XQuartz installed for X11 graphics to work.

Once connected, the console prompt should show something like `username@rackham1$`. This means you are on the login node. To view current jobs on queue, use `squeue -U username`. This should show a node number like `r120` etc if you have booked. If you have no bookings, see booking section below.

If you have a booked job, interactive job or any job, you can ssh on to that.

```{sh,block.title=TRUE,eval=FALSE}
ssh -Y username@r120
```

Your prompt changes to `username@r120$`. Now you are ready to start running heavy computations.

### Booking resources

The code below is valid to run at the start of the day. If you are running it in the middle of a day, you need to decrease the time (`-t`). Do not run this twice and also make sure you are not running computations on a login node.

Book resources for RNA-Seq day 2.

```{r,comment="",block.title=TRUE,class.output="sh",echo=FALSE}
cat(paste0("salloc -A ",id_project," -t 08:00:00 -p core -n 8 --reservation=",id_reservation_2))
```

Book resources for RNA-Seq day 3.

```{r,comment="",block.title=TRUE,class.output="sh",echo=FALSE}
cat(paste0("salloc -A ",id_project," -t 08:00:00 -p core -n 8 --reservation=",id_reservation_3))
```

<i class="fas fa-lightbulb"></i> Note that booking resources is only relevant for this course and not something you need to use when you regularly work with UPPMAX.

## Quality control

Before doing any other analysis on mapped RNA-seq reads it is always important to do quality control of your mapped reads and that you do not have any obvious errors in your RNA-seq data.

[<i class="fas fa-link"></i> <span class="large">Quality control</span>](lab_qc.html)

## Mapping

This section contains information on how to map reads to a reference genome using splice-aware aligner STAR and HISAT2.

[<i class="fas fa-link"></i> <span class="large">Mapping reads using STAR</span>](lab_mapping.html)

## IGV

Mapped reads in BAM files are visualised using integrated genome viewer.

[<i class="fas fa-link"></i> <span class="large">Using IGV</span>](lab_igv.html)

## Quantification

Gene counts are quantified from BAM files using featureCounts.

[<i class="fas fa-link"></i> <span class="large">Quantification</span>](lab_quantification.html)

## Differential gene expression

We find genes that are differentially expressed between our time points.

[<i class="fas fa-link"></i> <span class="large">DGE using DEseq2</span>](lab_dge.html)  

## Functional analysis

We will perform functional analysis on the differentially expressed genes to place them into a function context and possibly explain the biological consequences of DE. Methods covered are GSA (Gene set analysis) and GSEA (Gene set enrichment analysis).

[<i class="fas fa-link"></i> <span class="large">Functional analysis</span>](lab_functional.html)

# Bonus labs

## Exploratory data analyses

This section dives deeper into exploratory analyses PCA and hierarchical clustering.

[<i class="fas fa-link"></i> <span class="large">PCA & Hierarchical clustering</span>](lab_eda.html)

## Pseudoaligners

Kallisto uses FastQ reads and a reference transcriptome (cDNA+ncRNA) to quantify transcripts using rapid pseudo-alignment along with bootstrap replicates to assess quantification inaccuracy. Kallisto is significantly faster than STAR or HISAT2 and has a small memory footprint. Differential gene expression is carried out using Sleuth which utilises bootstrap replicates.

[<i class="fas fa-link"></i> <span class="large">Mapping and quantification using Kallisto, DGE using Sleuth</span>](lab_kallisto.html)  

## small RNA analyses

RNA-seq differential analyses workflow on microRNAs from Fruit fly.

[<i class="fas fa-link"></i> <span class="large">Small RNA-seq analyses</span>](lab_smallrna.html)

## Assembly & annotation

Raw sequencing short reads are assembled into transcripts using two approaches. Genome-guided assembly using HiSat2 and StringTie. De-novo transcriptome assembly using Trinity. Assembled transcriptomes are functionally annotated to identify genes.

[<i class="fas fa-link"></i> <span class="large">Reference-guided assembly using StringTie</span>](lab_assembly.html)  
[<i class="fas fa-link"></i> <span class="large">De-novo assembly using Trinity</span>](lab_assembly.html)  
[<i class="fas fa-link"></i> <span class="large">Transcriptome annotation</span>](lab_annotation.html)

**End of document**
