<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>DGE Workflow</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<link id="font-awesome-1-attachment" rel="attachment" href="site_libs/font-awesome-5.1.0/fonts/fontawesome-webfont.ttf"/>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="assets/lab.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
#TOC {
  display: none !important;
}
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img src="assets/logo.svg" id="logo" style="height:30px;margin:0;"/></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="home_schedule.html">Schedule</a>
</li>
<li>
  <a href="home_content.html">Content</a>
</li>
<li>
  <a href="home_precourse.html">Precourse</a>
</li>
<li>
  <a href="home_info.html">Info</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">DGE Workflow</h1>
<h3 class="subtitle">Workshop on RNA-Seq</h3>

</div>


<!-- rmd lab header -->
<p><br></p>
<div class="boxy boxy-exclamation">
<p>Create a directory named <code>data</code> in your current working directory for input and output files.</p>
</div>
<p>Load R packages.</p>
<pre class="r"><code># data handling
library(dplyr)
#library(tidyr)
#library(stringr)

# plotting
library(ggplot2)
library(gridExtra) # combining ggplots
library(biomaRt) # annotation
library(DESeq2) # rna-seq
library(edgeR) # rna-seq

# source download function
source(&quot;https://raw.githubusercontent.com/NBISweden/workshop-RNAseq/master/assets/scripts.R&quot;)</code></pre>
<div id="data-preprocessing" class="section level1">
<h1><span class="header-section-number">1</span> Data preprocessing</h1>
<p>Data preprocessing is done in R. First, we read in the count table.</p>
<pre class="r"><code>download_data(&quot;data/counts_raw.txt&quot;)
cr &lt;- read.table(&quot;./data/counts_raw.txt&quot;,header=TRUE)
head(cr)
str(cr)</code></pre>
<pre><code>##                 Sample_1 Sample_2 Sample_3 Sample_4 Sample_5 Sample_6 Sample_7
## ENSG00000000003      321      303      204      492      455      359      376
## ENSG00000000005        0        0        0        0        0        0        0
## ENSG00000000419      696      660      472      951      963      689      706
## ENSG00000000457       59       54       44      109       73       66       60
## ENSG00000000460      399      405      236      445      454      374      316
## ENSG00000000938        0        0        0        0        0        1        0
##                 Sample_8 Sample_9 Sample_10 Sample_11 Sample_12
## ENSG00000000003      523      450       950       760      1436
## ENSG00000000005        0        0         0         0         0
## ENSG00000000419     1041      796      1036       789      1413
## ENSG00000000457      125       74       108       115       174
## ENSG00000000460      505      398       141       168       259
## ENSG00000000938        0        0         1         0         0
## &#39;data.frame&#39;:    59573 obs. of  12 variables:
##  $ Sample_1 : int  321 0 696 59 399 0 0 844 535 493 ...
##  $ Sample_2 : int  303 0 660 54 405 0 0 808 458 444 ...
##  $ Sample_3 : int  204 0 472 44 236 0 1 495 290 308 ...
##  $ Sample_4 : int  492 0 951 109 445 0 2 1065 722 511 ...
##  $ Sample_5 : int  455 0 963 73 454 0 0 1037 693 523 ...
##  $ Sample_6 : int  359 0 689 66 374 1 0 839 574 350 ...
##  $ Sample_7 : int  376 0 706 60 316 0 1 723 406 370 ...
##  $ Sample_8 : int  523 0 1041 125 505 0 0 1266 623 579 ...
##  $ Sample_9 : int  450 0 796 74 398 0 0 984 545 497 ...
##  $ Sample_10: int  950 0 1036 108 141 1 0 1044 487 637 ...
##  $ Sample_11: int  760 0 789 115 168 0 1 877 446 548 ...
##  $ Sample_12: int  1436 0 1413 174 259 0 1 1428 648 878 ...</code></pre>
<p>The count data shows read counts across samples and genes. The columns denote samples and rows denote genes.</p>
<p>Read in the metadata. Each row corresponds to a sample. The sample names can be added as row names.</p>
<pre class="r"><code>download_data(&quot;data/metadata_raw.txt&quot;)
mr &lt;- read.csv2(&quot;./data/metadata_raw.txt&quot;,header=TRUE,stringsAsFactors=F)
rownames(mr) &lt;- mr$Sample_ID
head(mr)
str(mr)</code></pre>
<pre><code>##          Sample_ID Sample_Name Time Replicate Cell
## Sample_1  Sample_1        t0_A   t0         A A431
## Sample_2  Sample_2        t0_B   t0         B A431
## Sample_3  Sample_3        t0_C   t0         C A431
## Sample_4  Sample_4        t2_A   t2         A A431
## Sample_5  Sample_5        t2_B   t2         B A431
## Sample_6  Sample_6        t2_C   t2         C A431
## &#39;data.frame&#39;:    12 obs. of  5 variables:
##  $ Sample_ID  : chr  &quot;Sample_1&quot; &quot;Sample_2&quot; &quot;Sample_3&quot; &quot;Sample_4&quot; ...
##  $ Sample_Name: chr  &quot;t0_A&quot; &quot;t0_B&quot; &quot;t0_C&quot; &quot;t2_A&quot; ...
##  $ Time       : chr  &quot;t0&quot; &quot;t0&quot; &quot;t0&quot; &quot;t2&quot; ...
##  $ Replicate  : chr  &quot;A&quot; &quot;B&quot; &quot;C&quot; &quot;A&quot; ...
##  $ Cell       : chr  &quot;A431&quot; &quot;A431&quot; &quot;A431&quot; &quot;A431&quot; ...</code></pre>
<p>The metadata columns are sample name and time points. It is a good idea to check that the number of columns of data match the number of rows of metadata. The column names of data must also match the row names of metadata.</p>
<pre class="r"><code>all.equal(colnames(cr),rownames(mr))</code></pre>
<pre><code>## [1] TRUE</code></pre>
<p>Let’s create a boxplot to visualise the distribution of counts.</p>
<pre class="r"><code>boxplot(log10(as.matrix(cr)+1),ylab=expression(&#39;Log&#39;[10]~&#39;Read counts&#39;),las=2,
        main=&quot;Raw data&quot;)</code></pre>
<p><img src="lab_dge_files/figure-html/unnamed-chunk-6-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
<p>The median values are zero across all samples. Around half the values in each sample are zeros. This data set would benefit from a low count filtering.</p>
<p>We can check if any samples need to be discarded based on the number of genes detected. We create a barplot of genes detected across samples.</p>
<pre class="r"><code>{barplot(colSums(cr&gt;3),ylab=&quot;Number of detected genes&quot;,las=2)
abline(h=median(colSums(cr&gt;3)))}</code></pre>
<p><img src="lab_dge_files/figure-html/unnamed-chunk-7-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
<p>On average, about 15000 genes are detected. All samples are more or less close to the average. None of the samples look bad enough to be removed.</p>
<p><i class="fas fa-comments"></i> What does <code>cr&gt;3</code> do? Why did we use 3? Is it better than using <code>cr&gt;0</code>?</p>
<p>And we can create a similar plot for detection rate across genes.</p>
<pre class="r"><code>{barplot(rowSums(cr&gt;3),xlab=&quot;Genes&quot;,ylab=&quot;Number of samples&quot;,names.arg=&quot;&quot;)
abline(h=median(rowSums(cr&gt;3)),col=&quot;red&quot;)}</code></pre>
<p><img src="lab_dge_files/figure-html/unnamed-chunk-8-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
<p>This is harder to see. It’s perhaps easier to plot a histogram.</p>
<pre class="r"><code>hist(rowSums(cr&gt;3))</code></pre>
<p><img src="lab_dge_files/figure-html/unnamed-chunk-9-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
<p>There are a lot of genes that are not expressed (ie; zero on x-axis). These can be removed completely. We are really interested in the peak on the right. These are genes that are expressed in all samples. We don’t want to be too restrictive, so we will choose to keep genes that are expressed in atleast 3 samples.</p>
<p>Below, rather than using zero as the minimum value for detection, we used minimum of 1 count-per-million (CPM). This ignores noisy background counts. And we want to keep genes that have minimum 1 CPM across 2 samples (since each of test groups consist of 3 samples).</p>
<pre class="r"><code># remove genes with low counts
keepgenes &lt;- rowSums(edgeR::cpm(cr)&gt;1) &gt;= 2
cf &lt;- cr[keepgenes,]</code></pre>
<p><i class="fas fa-comments"></i> How would the results change if we used total number of samples (ie; 12 for this dataset) in the code above?</p>
<p>Distribution of the filtered counts looks like below. Compare this to the previous boxplot above.</p>
<pre class="r"><code>boxplot(log10(as.matrix(cf)+1),ylab=expression(&#39;Log&#39;[10]~&#39;Read counts&#39;),las=2,
        main=&quot;Filtered data&quot;)</code></pre>
<p><img src="lab_dge_files/figure-html/unnamed-chunk-11-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
<p>The missingness in the data set is reduced. The filtering process has removed 44995 genes with low counts.</p>
<p>Since no samples were discarded, the metadata file will remain the same. And we can check that the labels are the same order in counts and metadata.</p>
<pre class="r"><code>all.equal(colnames(cf),rownames(mr))</code></pre>
<pre><code>## [1] TRUE</code></pre>
<p>At this point, we can save the filtered data.</p>
<pre class="r"><code>write.table(cf,&quot;./data/counts_filtered.txt&quot;,col.names=T,quote=F,sep=&quot;\t&quot;,dec=&quot;.&quot;)</code></pre>
</div>
<div id="normalisation" class="section level1">
<h1><span class="header-section-number">2</span> Normalisation</h1>
<p>The raw count data needs to be corrected for various biases before statistical inference. If the data set is to be used in an R package for differential gene expression such as <strong>DESeq2</strong>, <strong>edgeR</strong> or <strong>Limma</strong>, the raw data must be used directly. This is because, these packages handle the correction and transformation internally.</p>
<div id="cpmtpm" class="section level2">
<h2><span class="header-section-number">2.1</span> CPM/TPM</h2>
<p>For analysis other than DGE, the data set must be corrected before use. The most basic correction required is sequencing depth. This is achieved using rescaling the counts to counts per million.</p>
<p>We read in out filtered count data and metadata. We can use the function <code>cpm()</code> from R package <strong>edgeR</strong> for this.</p>
<pre class="r"><code>download_data(&quot;data/counts_filtered.txt&quot;)
cf &lt;- read.table(&quot;data/counts_filtered.txt&quot;,header=TRUE)

download_data(&quot;data/metadata_raw.csv&quot;)
mr &lt;- read.csv2(&quot;data/metadata_raw.csv&quot;,header=TRUE,stringsAsFactors=F)
rownames(mr) &lt;- mr$Sample_ID</code></pre>
<pre class="r"><code>cc &lt;- edgeR::cpm(cf)
boxplot(log10(as.matrix(cc)+1),ylab=expression(&#39;Log&#39;[10]~&#39;Read counts&#39;),las=2,main=&quot;CPM&quot;)</code></pre>
<p><img src="lab_dge_files/figure-html/unnamed-chunk-15-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
<p>But, CPM data has some drawbacks. It is not suitable for within-sample comparisons. The total number of reads per sample varies from sample to sample. This also makes it harder to compare one experiment to another. In addition, gene length is not controlled for in this correction. RPKM/FPKM normalisations correct for gene length, but they are not recommended because they are not comparable between samples.</p>
<p>A better correction method that resolves these issues is TPM (transcripts-per-million). The code for computing TPM is simple.</p>
<pre class="r"><code>#&#39; @title Compute TPM from a read count matrix
#&#39; @param counts A numeric data.frame of read counts with samples (columns) and genes (rows).
#&#39; @param len A vector of gene cds length equal to number of rows of dfr.
#&#39;
#&#39; https://support.bioconductor.org/p/91218/
#&#39;
tpm &lt;- function(counts,len) {
  x &lt;- counts/len
  return(t(t(x)*1e6/colSums(x)))
}</code></pre>
<p>We read in the annotation data, remove duplicated ensembl IDs and compute gene lengths.</p>
<pre class="r"><code>download_data(&quot;data/human_genes.txt&quot;)
g &lt;- read.delim(&quot;data/human_genes.txt&quot;,header=T,stringsAsFactors=F)
g &lt;- g[!duplicated(g$ensembl_gene_id),]
g$len &lt;- g$end_position-g$start_position
rownames(g) &lt;- g$ensembl_gene_id</code></pre>
<p>Next, we find shared genes between count data and annotation data and match their order.</p>
<pre class="r"><code>igenes &lt;- intersect(rownames(cf),g$ensembl_gene_id)
g1 &lt;- g[igenes,]
cf1 &lt;- cf[igenes,]
all.equal(rownames(cf1),g1$ensembl_gene_id)</code></pre>
<pre><code>## [1] TRUE</code></pre>
<pre class="r"><code>ct &lt;- tpm(cf1,g1$len)
boxplot(log10(as.matrix(ct)+1),ylab=expression(&#39;Log&#39;[10]~&#39;Read counts&#39;),las=2,main=&quot;TPM&quot;)</code></pre>
<p><img src="lab_dge_files/figure-html/unnamed-chunk-19-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
<p>This is the distribution of TPM counts.</p>
</div>
<div id="deseq2" class="section level2">
<h2><span class="header-section-number">2.2</span> DESeq2</h2>
<p>DESeq2 internally corrects counts for sequencing depth and RNA compositional bias using <strong>Median of ratios</strong> method. The details of this method are described further below under DGE size factors. To run this method, we create a DESeq2 object using the count data and metadata.</p>
<pre class="r"><code>library(DESeq2)
mr$Time &lt;- factor(mr$Time)
d &lt;- DESeqDataSetFromMatrix(countData=cf,colData=mr,design=~Time)
d &lt;- DESeq2::estimateSizeFactors(d,type=&quot;ratio&quot;)
cd &lt;- counts(d,normalized=TRUE)
saveRDS(cd,&quot;data/counts_normalised_deseq2.Rds&quot;)</code></pre>
<pre class="r"><code>cd &lt;- readRDS(&quot;data/counts_normalised_deseq2.Rds&quot;)
boxplot(log10(as.matrix(cd)+1),ylab=expression(&#39;Log&#39;[10]~&#39;Read counts&#39;),las=2,main=&quot;DESeq2&quot;)</code></pre>
<p><img src="lab_dge_files/figure-html/unnamed-chunk-21-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</div>
<div id="vst" class="section level2">
<h2><span class="header-section-number">2.3</span> VST</h2>
<p>For the purpose of exploratory analysis such as MDS, PCA, clustering etc, VST (variance-stabilizing-transformation) is recommended. VST is also run using DESeq2. As in the previous step, a DESeq2 object is created.</p>
<pre class="r"><code>library(DESeq2)
mr$Time &lt;- factor(mr$Time)
d &lt;- DESeqDataSetFromMatrix(countData=cf,colData=mr,design=~Time)
d &lt;- DESeq2::estimateSizeFactors(d,type=&quot;ratio&quot;)
d &lt;- DESeq2::estimateDispersions(d)
cv &lt;- as.data.frame(assay(varianceStabilizingTransformation(d,blind=T)),check.names=F)

if(!file.exists(&quot;./data/counts_vst.txt&quot;)) {
  write.table(cv,&quot;./data/counts_vst.txt&quot;,sep=&quot;\t&quot;,dec=&quot;.&quot;,quote=FALSE)
}

boxplot(log10(as.matrix(cv)+1),ylab=expression(&#39;Log&#39;[10]~&#39;Read counts&#39;),las=2,main=&quot;VST&quot;)</code></pre>
<p><img src="lab_dge_files/figure-html/unnamed-chunk-22-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
<p>The effect of VST transformation can be clearly seen in a mean vs variance plot.</p>
<pre class="r"><code>rowVar &lt;- function(x) apply(x,1,var)
par(mfrow=c(2,2))
plot(log10(rowMeans(cf)+1),log10(rowVar(cf)+1),xlab=expression(&#39;Log&#39;[10]~&#39;Mean count&#39;),ylab=expression(&#39;Log&#39;[10]~&#39;Variance&#39;),main=&quot;Filtered&quot;)
plot(log10(rowMeans(ct)+1),log10(rowVar(ct)+1),xlab=expression(&#39;Log&#39;[10]~&#39;Mean count&#39;),ylab=expression(&#39;Log&#39;[10]~&#39;Variance&#39;),main=&quot;TPM&quot;)
plot(log10(rowMeans(cd)+1),log10(rowVar(cd)+1),xlab=expression(&#39;Log&#39;[10]~&#39;Mean count&#39;),ylab=expression(&#39;Log&#39;[10]~&#39;Variance&#39;),main=&quot;DESeq2&quot;)
plot(rowMeans(cv),rowVar(cv),xlab=&#39;Mean count&#39;,ylab=&#39;Variance&#39;,main=&quot;VST&quot;)
par(mfrow=c(1,1))</code></pre>
<p><img src="lab_dge_files/figure-html/unnamed-chunk-23-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
<p>For RNA-seq data, as the mean count value increases, the variance increases. There is a strong almost linear relationship as seen in the figures. The statistical methods such as PCA expects similar variance across the range of mean values. If not, the higher variance genes will contribute more than the lower variance genes. Such data is said to be heteroscedastic and needs to be corrected. Correction using log transformation (with pseudocount) now gives inflates the contribution of the low variance genes. To obtain similar variance across the whole range of mean values, DESeq2 offers two methods VST (variance stabilising transformation) and RLOG (regularised log transformation).</p>
<p>As the name suggests, VST transformation stabilizes variance across the whole range of count values. VST is recommended for clustering or visualisation. It is not intended for differential gene expression. If the size factors vary dramatically between samples, then RLOG transformation is recommended.</p>
<p>Finally, we can compare all of the various transformations in a single plot.</p>
<pre class="r"><code>par(mfrow=c(1,4))
boxplot(log10(as.matrix(cf)+1),ylab=expression(&#39;Log&#39;[10]~&#39;Read counts&#39;),las=2,main=&quot;Filtered&quot;)
boxplot(log10(as.matrix(ct)+1),ylab=expression(&#39;Log&#39;[10]~&#39;Read counts&#39;),las=2,main=&quot;TPM&quot;)
boxplot(log10(as.matrix(cd)+1),ylab=expression(&#39;Log&#39;[10]~&#39;Read counts&#39;),las=2,main=&quot;DESeq2&quot;)
boxplot(as.matrix(cv),ylab=&#39;Read counts&#39;,las=2,main=&quot;VST&quot;)
par(mfrow=c(1,1))</code></pre>
<p><img src="lab_dge_files/figure-html/unnamed-chunk-24-1.svg" width="864" style="display: block; margin: auto auto auto 0;" /></p>
<p><i class="fas fa-comments"></i> Would it be possible to have one perfect normalisation method for all types of analyses? Is there any drawback to using gene length corrected counts in differential gene expression analyses?</p>
</div>
</div>
<div id="exploratory" class="section level1">
<h1><span class="header-section-number">3</span> Exploratory</h1>
<p>In this section, we are strictly not running any quantitative analyses or statistical test. This is a QC of the counts. We are running correlation to check similarity of gene counts between samples. We are running distance based clustering and PCA to estimate similarity between samples. This should give us an idea of the variation within your sample groups and detect possible outliers or mis-labelled samples.</p>
<p>We will use the variance stabilized counts (VST) from the previous step for all exploratory analyses. Otherwise, the files below can also be downloaded.</p>
<pre class="r"><code>download_data(&quot;data/counts_vst.txt&quot;)
cv &lt;- read.table(&quot;data/counts_vst.txt&quot;,header=TRUE)

download_data(&quot;data/metadata_raw.csv&quot;)
mr &lt;- read.csv2(&quot;data/metadata_raw.csv&quot;,header=TRUE,stringsAsFactors=F)
rownames(mr) &lt;- mr$Sample_ID</code></pre>
<div id="correlation" class="section level2">
<h2><span class="header-section-number">3.1</span> Correlation</h2>
<p>It is a good idea to start by checking correlation between samples. RNA-Seq samples generally have very high correlation (R^2 &gt; 0.9). R^2 values below 0.8 may be an indication of an outlier sample. Depending on other QC metric, these low correlation samples may be discarded from analyses.</p>
<p>We use the function <code>cor()</code> for computing sample-to-sample Spearman correlation. Note that the input matrix has genes as rows and samples as columns. This generates a sample-to-sample pairwise correlation matrix. This matrix can be plotted as a heatmap using the <code>pheatmap()</code> function from the <strong>pheatmap</strong> R package.</p>
<pre class="r"><code>dmat &lt;- as.matrix(cor(cv,method=&quot;spearman&quot;))

library(pheatmap)
pheatmap(dmat,border_color=NA,annotation_col=mr[,&quot;Time&quot;,drop=F],
         annotation_row=mr[,&quot;Time&quot;,drop=F],annotation_legend=T)</code></pre>
<p><img src="lab_dge_files/figure-html/unnamed-chunk-26-1.svg" width="576" style="display: block; margin: auto auto auto 0;" /></p>
<p>In the matrix, red colour denotes higher correlation (higher similarity) and blue denotes lower correlation (lower similarity). <code>pheatmap()</code> also hierarchically clusters rows and columns based on correlation values. The dendrograms show how samples cluster. Annotation colours denote <strong>Time</strong> groups. Notice that samples group by <strong>Time</strong>.</p>
</div>
<div id="pca" class="section level2">
<h2><span class="header-section-number">3.2</span> PCA</h2>
<p>To run PCA, we use the R function <code>prcomp()</code>. It takes in a matrix where samples are rows and variables are columns. Therefore, we transpose our count matrix using the function <code>t()</code>. If we do not transpose, then PCA is run on the genes rather than the samples. The next line of code plots the variance explained by the top PCs.</p>
<pre class="r"><code>pcaobj &lt;- prcomp(x=t(cv))
{barplot(round(pcaobj$sdev^2/sum(pcaobj$sdev^2)*100,2),las=2,
        names.arg=colnames(pcaobj$x),ylab=&quot;% Variance explained&quot;,
        xlab=&quot;PCA principal components&quot;)
abline(h=2, lty=2)}</code></pre>
<p><img src="lab_dge_files/figure-html/unnamed-chunk-27-1.svg" width="480" style="display: block; margin: auto auto auto 0;" /></p>
<p>The first two principal components in total explain 85% (75%+10%) of the variance in the data set. This means a scatterplot of PC1 vs PC2 can help to visualise the most important trend in the data. Then we merge the rotated data with the metadata and plot a scatterplot coloured by our variable of interest (Time).</p>
<pre class="r"><code>pcamat1 &lt;- as.data.frame(pcaobj$x)
pcamat2 &lt;- merge(pcamat1,mr,by=0)

ggplot(pcamat2,aes(PC1,PC2,colour=Time))+
  geom_point()+
  theme_bw()</code></pre>
<p><img src="lab_dge_files/figure-html/unnamed-chunk-28-1.svg" width="432" style="display: block; margin: auto auto auto 0;" /></p>
<p>Samples cluster by the <strong>Time</strong> variable as expected.</p>
<p>Sometimes the first two PCs may not be the ones that will best separate the sample groups, so it is a good idea to look at more PCs.</p>
<pre class="r"><code>p1 &lt;- ggplot(pcamat2,aes(PC1,PC3,colour=Time))+
  geom_point()+
  theme_bw()
p2 &lt;- ggplot(pcamat2,aes(PC2,PC3,colour=Time))+
  geom_point()+
  theme_bw()
gridExtra::grid.arrange(p1,p2,nrow=1,ncol=2)</code></pre>
<p><img src="lab_dge_files/figure-html/unnamed-chunk-29-1.svg" width="768" style="display: block; margin: auto auto auto 0;" /></p>
<p>An alternative way to create a PCA plot is directly from the DESeq2 object using the DESeq2 function <code>plotPCA()</code>.</p>
<pre class="r"><code>plotPCA(varianceStabilizingTransformation(d),intgroup=&quot;Time&quot;)</code></pre>
<p><img src="lab_dge_files/figure-html/unnamed-chunk-30-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</div>
<div id="clustering" class="section level2">
<h2><span class="header-section-number">3.3</span> Clustering</h2>
<p>For clustering, we create a sample-to-sample pairwise distance matrix (here we use euclidean distance). The rows and columns of this matrix is then hierarchically clustered. We use the function <code>dist()</code> to compute the distance. Note that for a sample-to-sample matrix, the rows need to be samples and columns should be genes. Therefore, we use the function <code>t()</code> to transpose our VST normalised count matrix.</p>
<pre class="r"><code>dmat &lt;- as.matrix(dist(t(cv)))

library(pheatmap)
pheatmap(dmat,border_color=NA,annotation_col=mr[,&quot;Time&quot;,drop=F],
         annotation_row=mr[,&quot;Time&quot;,drop=F],annotation_legend=T)</code></pre>
<p><img src="lab_dge_files/figure-html/unnamed-chunk-31-1.svg" width="576" style="display: block; margin: auto auto auto 0;" /></p>
<p>Hierarchically clustered sample-to-sample euclidean distance matrix. Larger distances mean lower similarity and vice-versa. In the matrix, red colour denotes larger distance (lower similarity) and blue denotes small distance (higher similarity). Annotation colours denote <strong>Time</strong> groups. The dendrogram helps to visualise sample clustering. Notice that samples group by <strong>Time</strong>.</p>
<p><i class="fas fa-clipboard-list"></i> Try to run the PCA using one of the other normalisation methods, say logCPM and/or DESeq2 normalised counts.
<i class="fas fa-comments"></i> How different are the results? How do samples group? Can these differences be explained?</p>
</div>
</div>
<div id="dge" class="section level1">
<h1><span class="header-section-number">4</span> DGE</h1>
<p>For differential gene expression, we use the <strong>DESeq2</strong> package. We use the raw filtered counts and metadata.</p>
<pre class="r"><code>download_data(&quot;data/counts_filtered.txt&quot;)
cf &lt;- read.table(&quot;data/counts_filtered.txt&quot;,header=TRUE)

download_data(&quot;data/metadata_raw.csv&quot;)
mr &lt;- read.csv2(&quot;data/metadata_raw.csv&quot;,header=TRUE,stringsAsFactors=F)
rownames(mr) &lt;- mr$Sample_ID</code></pre>
<p>The data is converted to a DESeq2 object. The GLM model we use is simple since we only have one variable of interest <code>~Time</code>.</p>
<p>If we had other covariates to control for, we would add them to the model like so <code>~var+Time</code>. The variable of interest appears in the end. This model means find differentially expressed genes between groups under ‘time’ variable while controlling for the effect of ‘var’. Similarily, batch effects can be controlled by specifying them in the model <code>~batch+Time</code>.</p>
<pre class="r"><code>library(DESeq2)
mr$Time &lt;- factor(mr$Time)
d &lt;- DESeqDataSetFromMatrix(countData=cf,colData=mr,design=~Time)</code></pre>
<div id="size-factors" class="section level2">
<h2><span class="header-section-number">4.1</span> Size factors</h2>
<p>The first step is estimating size factors. The data is normalised for sequencing depth and compositional bias as done for the VST step. DESeq2 uses a method called <em>median-of-ratios</em> for this step.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>d &lt;- DESeq2::estimateSizeFactors(d,type=&quot;ratio&quot;)</code></pre>
</div>
<div class="boxy boxy-optional">
<p><strong>Optional</strong></p>
<p>For those interested in the details of the <em>median-of-ratios</em> method, click below.</p>
<p>
&lt;button class=“btn btn-sm btn-primary btn-collapse btn-collapse-optional collapsed type=”button" data-toggle=“collapse” data-target=“#dge-size-factor” aria-expanded=“false” aria-controls=“dge-size-factor”&gt;
</button>
</p>
<div id="dge-size-factor" class="collapse">
<div class="card card-body">
<p>This is a step-by-step guide to computing normalisation factors (size factors) using the <em>median-of-ratios</em> method.</p>
<ol style="list-style-type: decimal">
<li>The geometric mean is computed across all samples for each gene.</li>
</ol>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>gm_mean = function(x, na.rm=TRUE){
  exp(sum(log(x[x &gt; 0]), na.rm=na.rm) / length(x))
}

gmean &lt;- apply(cf,1,gm_mean)
head(gmean)</code></pre>
<pre><code>## ENSG00000000003 ENSG00000000419 ENSG00000000457 ENSG00000000460 ENSG00000001036 
##       479.11181       820.01611        81.68888       319.29571       920.13961 
## ENSG00000001084 
##       520.55995</code></pre>
</div>
<ol start="2" style="list-style-type: decimal">
<li>Read counts for each gene is divided by the geometric mean of that gene to create a ratio.</li>
</ol>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>ratio &lt;- cf/gmean
head(ratio)[,1:5]</code></pre>
<pre><code>##                  Sample_1  Sample_2  Sample_3 Sample_4  Sample_5
## ENSG00000000003 0.6699898 0.6324202 0.4257879 1.026900 0.9496739
## ENSG00000000419 0.8487638 0.8048622 0.5755984 1.159733 1.1743672
## ENSG00000000457 0.7222525 0.6610447 0.5386290 1.334331 0.8936344
## ENSG00000000460 1.2496253 1.2684166 0.7391267 1.393692 1.4218794
## ENSG00000001036 0.9172521 0.8781276 0.5379618 1.157433 1.1270029
## ENSG00000001084 1.0277395 0.8798218 0.5570924 1.386968 1.3312588</code></pre>
</div>
<ol start="3" style="list-style-type: decimal">
<li>The median ratio for each sample (across all genes) is taken as the size factor for that sample.</li>
</ol>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>sf &lt;- apply(ratio,2,median)
sf</code></pre>
<pre><code>##  Sample_1  Sample_2  Sample_3  Sample_4  Sample_5  Sample_6  Sample_7  Sample_8 
## 0.8975253 0.8407580 0.5086763 1.1263621 1.0931051 0.8123073 0.7558324 1.1745210 
##  Sample_9 Sample_10 Sample_11 Sample_12 
## 1.0193067 1.3720949 1.2387549 1.8679810</code></pre>
</div>
<p>We can verify that these values are correct by comparing with size factors generated by DESeq2.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code># deseq2 size factors
sizeFactors(d)</code></pre>
<pre><code>##  Sample_1  Sample_2  Sample_3  Sample_4  Sample_5  Sample_6  Sample_7  Sample_8 
## 0.9003753 0.8437393 0.5106445 1.1276451 1.0941383 0.8133849 0.7553903 1.1744008 
##  Sample_9 Sample_10 Sample_11 Sample_12 
## 1.0189325 1.3642797 1.2325485 1.8555904</code></pre>
</div>
<p>If we plot the size factors for each sample against the total counts for each sample, we get the plot below. We can see that they correlate very well. Size factors are mostly correcting for total counts, ie; sequencing depth.</p>
<pre class="r"><code>plot(sizeFactors(d),colSums(cf),xlab=&quot;Size factors&quot;,ylab=&quot;Total counts&quot;)</code></pre>
<p><img src="lab_dge_files/figure-html/unnamed-chunk-39-1.svg" width="432" style="display: block; margin: auto auto auto 0;" /></p>
<p>The raw counts can then be divided by the size factor to yield normalised read counts.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code># custom
head(t(t(cf)/sf))[,1:5]
# deseq2
head(counts(d,normalized=TRUE))[,1:5]</code></pre>
<pre><code>##                  Sample_1  Sample_2  Sample_3  Sample_4  Sample_5
## ENSG00000000003 357.65009 360.38907 401.04088 436.80445 416.24541
## ENSG00000000419 775.46562 785.00590 927.89851 844.31105 880.97655
## ENSG00000000457  65.73631  64.22776  86.49901  96.77172  66.78223
## ENSG00000000460 444.55572 481.70816 463.94925 395.07720 415.33059
## ENSG00000001036 940.36349 961.03752 973.11390 945.52183 948.67361
## ENSG00000001084 596.08349 544.74652 570.10713 641.00166 633.97378
##                  Sample_1  Sample_2  Sample_3  Sample_4  Sample_5
## ENSG00000000003 356.51801 359.11565 399.49511 436.30747 415.85236
## ENSG00000000419 773.01102 782.23211 924.32203 843.35041 880.14467
## ENSG00000000457  65.52823  64.00081  86.16561  96.66161  66.71917
## ENSG00000000460 443.14856 480.00607 462.16101 394.62769 414.93840
## ENSG00000001036 937.38692 957.64173 969.36314 944.44605 947.77780
## ENSG00000001084 594.19669 542.82168 567.90972 640.27234 633.37514</code></pre>
</div>
</div>
</div>
</div>
<p><i class="fas fa-comments"></i> The function <code>estimateSizeFactors()</code> has options to set a custom <code>locfunc</code> other than the median. Why is this useful? What happens if you change it to “shorth”. Check out the help page for <code>estimateSizeFactors()</code>.</p>
</div>
<div id="gene-dispersion" class="section level2">
<h2><span class="header-section-number">4.2</span> Gene dispersion</h2>
<p>When it comes to comparing values between groups, some measure of variation is needed to estimate the variability in gene counts within groups. Dispersion is a measure of variation in a data set. Variance and standard deviation are not a good measure to estimate variability because it correlates with the mean.</p>
<div class="boxy boxy-optional">
<p><strong>Optional</strong></p>
<p>For some more discussion on dispersion, click below.</p>
<p>
&lt;button class=“btn btn-sm btn-primary btn-collapse btn-collapse-optional collapsed type=”button" data-toggle=“collapse” data-target=“#dge-dispersion” aria-expanded=“false” aria-controls=“dge-dispersion”&gt;
</button>
</p>
<div id="dge-dispersion" class="collapse">
<div class="card card-body">
<p>We can create a mean counts vs variance plot for all genes in our data set.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>dm &lt;- apply(cd,1,mean)
dv &lt;- apply(cd,1,var)

ggplot(data.frame(mean=log10(dm+1),var=log10(dv+1)),
       aes(mean,var))+
  geom_point(alpha=0.2)+
  geom_smooth(method=&quot;lm&quot;)+
  labs(x=expression(&#39;Log&#39;[10]~&#39;Mean counts&#39;),y=expression(&#39;Log&#39;[10]~&#39;Variance&#39;))+
  theme_bw()</code></pre>
<img src="lab_dge_files/figure-html/unnamed-chunk-41-1.svg" width="432" style="display: block; margin: auto auto auto 0;" />
</div>
<p>We see a mostly linear relationship on the log scale. The blue line denotes a linear fit. Genes that have larger read counts show higher variance. It’s hard to say which genes are more variable based on this alone. Therefore, variance is not a good measure to identify variation in read counts. A measure that controls for this mean-variance relationship is what we need.</p>
<p>One option is the coefficient of variation (CV).</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>cva &lt;- function(x) sd(x)/mean(x)
dc &lt;- apply(cd,1,cva)

ggplot(data.frame(mean=log10(dm+1),var=dc),
       aes(mean,var))+
  geom_point(alpha=0.2)+
  geom_smooth()+
  labs(x=expression(&#39;Log&#39;[10]~&#39;Mean counts&#39;),y=&quot;Coefficient of variation&quot;)+
  theme_bw()</code></pre>
<img src="lab_dge_files/figure-html/unnamed-chunk-42-1.svg" width="432" style="display: block; margin: auto auto auto 0;" />
</div>
<p>Now, we see that genes with lower counts have higher variability and genes with larger counts have lower variability. A measure like CV is taking the ratio of ‘variation’ to mean. <code>cv=sd(x)/mean(x)</code>.</p>
<p>This becomes even more apparent if we compute the CV and mean over replicates within sample groups (Time).</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>dx1 &lt;- data.frame(t0=apply(cd[,1:3],1,cva),t2=apply(cd[,4:6],1,cva),
            t6=apply(cd[,7:9],1,cva),t24=apply(cd[,10:12],1,cva))
dx1$gene &lt;- rownames(dx1)
dx1 &lt;- tidyr::gather(dx1,key=sample,value=cv,-gene)
rownames(dx1) &lt;- paste0(dx1$gene,&quot;-&quot;,dx1$sample)

dx2 &lt;- data.frame(t0=apply(cd[,1:3],1,mean),t2=apply(cd[,4:6],1,mean),
            t6=apply(cd[,7:9],1,mean),t24=apply(cd[,10:12],1,mean))
dx2$gene &lt;- rownames(dx2)
dx2 &lt;- tidyr::gather(dx2,key=sample,value=mean,-gene)
rownames(dx2) &lt;- paste0(dx2$gene,&quot;-&quot;,dx2$sample)

dx3 &lt;- merge(dx1,dx2,by=0)

ggplot(dx3,aes(x=log10(mean+1),y=cv))+
  geom_point(alpha=0.2)+
  geom_smooth()+
  facet_wrap(~sample.x)+
  labs(x=expression(&#39;Log&#39;[10]~&#39;Mean counts&#39;),y=&quot;Coefficient of variation&quot;)+
  theme_bw()</code></pre>
<img src="lab_dge_files/figure-html/unnamed-chunk-43-1.svg" width="480" style="display: block; margin: auto auto auto 0;" />
</div>
<p>We find that CV strongly declines with increasing counts. Genes with low counts show higher dispersion. For the sake of completeness, we can also plot the relationship between CV and variance for the same sample groups.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>dx1 &lt;- data.frame(t0=apply(cd[,1:3],1,cva),t2=apply(cd[,4:6],1,cva),
            t6=apply(cd[,7:9],1,cva),t24=apply(cd[,10:12],1,cva))
dx1$gene &lt;- rownames(dx1)
dx1 &lt;- tidyr::gather(dx1,key=sample,value=cv,-gene)
rownames(dx1) &lt;- paste0(dx1$gene,&quot;-&quot;,dx1$sample)

dx2 &lt;- data.frame(t0=apply(cd[,1:3],1,var),t2=apply(cd[,4:6],1,var),
            t6=apply(cd[,7:9],1,var),t24=apply(cd[,10:12],1,var))
dx2$gene &lt;- rownames(dx2)
dx2 &lt;- tidyr::gather(dx2,key=sample,value=var,-gene)
rownames(dx2) &lt;- paste0(dx2$gene,&quot;-&quot;,dx2$sample)

dx3 &lt;- merge(dx1,dx2,by=0)

ggplot(dx3,aes(x=log10(var+1),y=cv))+
  geom_point(alpha=0.2)+
  geom_smooth()+
  facet_wrap(~sample.x)+
  labs(x=expression(&#39;Log&#39;[10]~&#39;Variance&#39;),y=&quot;Coefficient of variation&quot;)+
  theme_bw()</code></pre>
<img src="lab_dge_files/figure-html/unnamed-chunk-44-1.svg" width="480" style="display: block; margin: auto auto auto 0;" />
</div>
</div>
</div>
</div>
<p>DESeq2 computes it’s own version of dispersion in a more robust manner taking into account low count values. The DESeq2 dispersion estimates are inversely related to the mean and directly related to variance. The dispersion estimate is a good measure of the variation in gene expression for a certain mean value.</p>
<p>Now, the variance or dispersion estimate for genes with low counts is unreliable when there are too few replicates. To overcome this, DESeq2 borrows information from other genes. DESeq2 assumes that genes with similar expression levels have similar dispersion values. Dispersion estimates are computed for each gene separately using maximum likelihood estimate. A curve is fitted to these gene-wise dispersion estimates. The gene-wise estimates are then ‘shrunk’ to the fitted curve.</p>
<p>Gene-wide dispersions, fitted curve and shrinkage can be visualised using the <code>plotDispEsts()</code> function.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>d &lt;- DESeq2::estimateDispersions(d)
plotDispEsts(d)</code></pre>
<img src="lab_dge_files/figure-html/unnamed-chunk-45-1.svg" width="480" style="display: block; margin: auto auto auto 0;" />
</div>
<p>The black points denote the maximum likelihood dispersion estimate for each gene. The red curve denote the fitted curve. The blue points denote the new gene dispersion estimates after shrunk towards the curve. The circled blue points denote estimates that are not shrunk as they are too far away from the curve. Thus, shrinkage method is important to reduce false positives in DGE analysis involving too few replicates.</p>
<p><i class='fas fa-lightbulb'></i> It is a good idea to visually check the dispersion shrinkage plot to verify that the modelling works for your data set.</p>
</div>
<div id="testing" class="section level2">
<h2><span class="header-section-number">4.3</span> Testing</h2>
<p>Overdispersion is the reason why RNA-Seq data is better modelled as negative-binomial distribution rather than poisson distribution. Poisson distribution has a mean = variance relationship, while negative-binomial distribution has a variance &gt; mean relationship. The last step in the DESeq2 workflow is fitting the Negative Binomial model for each gene and performing differential expression testing. This is based on the log fold change values computed on the corrected count estimates between groups.</p>
<p><code>logFC = log2 (corrected counts group A / corrected counts group B)</code></p>
<p>The most commonly used testing for comparing two groups in DESeq2 is the Walds’s test. The null hypothesis is that the groups are not different and <code>logFC=0</code>. The list of contrasts can be seen using <code>resultsNames()</code>. Then we can pick our comparisons of interest.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>dg &lt;- nbinomWaldTest(d)
resultsNames(dg)</code></pre>
<pre><code>## [1] &quot;Intercept&quot;      &quot;Time_t2_vs_t0&quot;  &quot;Time_t24_vs_t0&quot; &quot;Time_t6_vs_t0&quot;</code></pre>
</div>
<p>And we can get the result tables for the three different comparisons. The summary of the result object shows the number of genes that are differentially expressed with positive or negative fold-change and outliers.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>res1 &lt;- results(dg,name=&quot;Time_t2_vs_t0&quot;,alpha=0.05)
summary(res1)
res2 &lt;- results(dg,name=&quot;Time_t24_vs_t0&quot;,alpha=0.05)
summary(res2)
res3 &lt;- results(dg,name=&quot;Time_t6_vs_t0&quot;,alpha=0.05)
summary(res3)</code></pre>
<pre><code>## 
## out of 14578 with nonzero total read count
## adjusted p-value &lt; 0.05
## LFC &gt; 0 (up)       : 413, 2.8%
## LFC &lt; 0 (down)     : 696, 4.8%
## outliers [1]       : 0, 0%
## low counts [2]     : 2261, 16%
## (mean count &lt; 26)
## [1] see &#39;cooksCutoff&#39; argument of ?results
## [2] see &#39;independentFiltering&#39; argument of ?results
## 
## 
## out of 14578 with nonzero total read count
## adjusted p-value &lt; 0.05
## LFC &gt; 0 (up)       : 5066, 35%
## LFC &lt; 0 (down)     : 5093, 35%
## outliers [1]       : 0, 0%
## low counts [2]     : 0, 0%
## (mean count &lt; 3)
## [1] see &#39;cooksCutoff&#39; argument of ?results
## [2] see &#39;independentFiltering&#39; argument of ?results
## 
## 
## out of 14578 with nonzero total read count
## adjusted p-value &lt; 0.05
## LFC &gt; 0 (up)       : 2484, 17%
## LFC &lt; 0 (down)     : 2856, 20%
## outliers [1]       : 0, 0%
## low counts [2]     : 0, 0%
## (mean count &lt; 3)
## [1] see &#39;cooksCutoff&#39; argument of ?results
## [2] see &#39;independentFiltering&#39; argument of ?results</code></pre>
</div>
<p>You can also build up the comparison using <strong>contrasts</strong>. Contrats need the condition, level to compare and the reference level (base level). For example, <code>results(dg,contrast=c("Time","t2","t0"),alpha=0.05)</code>.</p>
<p>Note that in both cases above, <code>t0</code> is the reference level and other levels are compared to this. Therefore, a fold-change of 2 would mean that, a gene is 2 fold higher expressed in the test level compared to <code>t0</code>.</p>
<p>The <code>results()</code> function has many useful arguments. One can set a threshold on the logFC values using <code>lfcThreshold</code>. By default, no filtering is performed based on logFC values. Outliers are detected and p-values are set to NA automatically using <code>cooksCutoff</code>. <code>independentFiltering</code> remove genes with low counts.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>head(res1)</code></pre>
<pre><code>## log2 fold change (MLE): Time t2 vs t0 
## Wald test p-value: Time t2 vs t0 
## DataFrame with 6 rows and 6 columns
##                  baseMean log2FoldChange     lfcSE      stat    pvalue
##                 &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
## ENSG00000000003  490.0172      0.2206198 0.1127611  1.956524 0.0504034
## ENSG00000000419  817.7807      0.0592720 0.1014813  0.584068 0.5591746
## ENSG00000000457   82.0788      0.2077486 0.2204049  0.942577 0.3458972
## ENSG00000000460  356.0716     -0.1291864 0.1151392 -1.122002 0.2618616
## ENSG00000001036  919.6068      0.0288827 0.0851501  0.339198 0.7344609
## ENSG00000001084  529.5940      0.2119648 0.0929811  2.279655 0.0226281
##                      padj
##                 &lt;numeric&gt;
## ENSG00000000003  0.263505
## ENSG00000000419  0.830262
## ENSG00000000457  0.689946
## ENSG00000000460  0.612625
## ENSG00000001036  0.909639
## ENSG00000001084  0.159263</code></pre>
</div>
<p>The results table contains mean expression value (baseMean), log2 fold change (log2FoldChange), log2 fold change standard error (lfcSE), wald test statistic (stat), wald test p-value (pvalue) and BH adjusted p-value (padj) for each gene.</p>
<p>It is a good idea to look at the distribution on unadjusted p-values.</p>
<pre class="r"><code>hist(res1$pvalue[res1$baseMean&gt;1],main=&quot;res1 Pval distribution&quot;,xlab=&quot;P-values&quot;)</code></pre>
<p><img src="lab_dge_files/figure-html/unnamed-chunk-49-1.svg" width="480" style="display: block; margin: auto auto auto 0;" /></p>
<p>This is the kind of distribution to be expected when the p-values are “well-behaved”. For more explanation on p-value distributions, see <a href="http://varianceexplained.org/statistics/interpreting-pvalue-histogram/">here</a>. If this distribution is very different or strange, then it might indicate an underlying problem.</p>
<p><i class='fas fa-lightbulb'></i> Note that the results object is a <code>DESeqResults</code> class object and not a data.frame. It can be converted to a data.frame using <code>as.data.frame()</code> for exporting to a file.</p>
<p>We can filter the results table as needed.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code># all genes
nrow(as.data.frame(res1))
# only genes with padj &lt;0.05
nrow(filter(as.data.frame(res1),padj&lt;0.05))
# only genes with padj &lt;0.05 and an absolute fold change &gt;2
nrow(filter(as.data.frame(res1),padj&lt;0.05,abs(log2FoldChange)&gt;2))</code></pre>
<pre><code>## [1] 14578
## [1] 1109
## [1] 11</code></pre>
</div>
<p>Note that manually filtering by <strong>log2FoldChange</strong> on the results table is not the same as setting the <code>lfcThreshold</code> argument in the <code>results()</code> function.</p>
<p>Finally, we can add additional information to our results to make the interpretation and downstream analyses easier. We read in the gene info downloaded through biomaRt.</p>
<pre class="r"><code># convert res to data.frame
df_res &lt;- as.data.frame(res1)
# add row names as a new column
df_res$ensembl_gene_id &lt;- rownames(df_res)

# read genes info
download_data(&quot;data/human_genes.txt&quot;)
hg &lt;- read.delim(&quot;data/human_genes.txt&quot;,header=T,sep=&quot;\t&quot;,stringsAsFactors=F)
hg &lt;- hg[!duplicated(hg$ensembl_gene_id),]

# merge dataframes
df_res1 &lt;- merge(df_res,hg,by=&quot;ensembl_gene_id&quot;)
head(df_res1)</code></pre>
<pre><code>##   ensembl_gene_id  baseMean log2FoldChange      lfcSE       stat     pvalue
## 1 ENSG00000000003 490.01721     0.22061981 0.11276108  1.9565244 0.05040341
## 2 ENSG00000000419 817.78066     0.05927197 0.10148128  0.5840680 0.55917460
## 3 ENSG00000000457  82.07877     0.20774862 0.22040490  0.9425772 0.34589722
## 4 ENSG00000000460 356.07160    -0.12918638 0.11513918 -1.1220019 0.26186163
## 5 ENSG00000001036 919.60675     0.02888269 0.08515007  0.3391975 0.73446094
## 6 ENSG00000001084 529.59397     0.21196477 0.09298106  2.2796555 0.02262813
##        padj entrezgene external_gene_name chromosome_name start_position
## 1 0.2635055       7105             TSPAN6               X      100627109
## 2 0.8302623       8813               DPM1              20       50934867
## 3 0.6899459      57147              SCYL3               1      169849631
## 4 0.6126246      55732           C1orf112               1      169662007
## 5 0.9096386       2519              FUCA2               6      143494811
## 6 0.1592633       2729               GCLC               6       53497341
##   end_position strand   gene_biotype
## 1    100639991     -1 protein_coding
## 2     50958555     -1 protein_coding
## 3    169894267     -1 protein_coding
## 4    169854080      1 protein_coding
## 5    143511690     -1 protein_coding
## 6     53616970     -1 protein_coding
##                                                                                      description
## 1                                              tetraspanin 6 [Source:HGNC Symbol;Acc:HGNC:11858]
## 2 dolichyl-phosphate mannosyltransferase subunit 1, catalytic [Source:HGNC Symbol;Acc:HGNC:3005]
## 3                                   SCY1 like pseudokinase 3 [Source:HGNC Symbol;Acc:HGNC:19285]
## 4                        chromosome 1 open reading frame 112 [Source:HGNC Symbol;Acc:HGNC:25565]
## 5                                        alpha-L-fucosidase 2 [Source:HGNC Symbol;Acc:HGNC:4008]
## 6                 glutamate-cysteine ligase catalytic subunit [Source:HGNC Symbol;Acc:HGNC:4311]</code></pre>
<p>Now we have annotated our DEG list with information such as gene names, genomic coordinates, biotype etc.</p>
<div id="lfcshrink" class="section level3">
<h3><span class="header-section-number">4.3.1</span> lfcShrink</h3>
<p>This is an extra step to generate more accurate log2 fold changes. This step corrects the log2 fold changes for genes with high dispersion.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>lres1 &lt;- lfcShrink(dg,coef=&quot;Time_t2_vs_t0&quot;,res=res1,type=&quot;normal&quot;)
summary(lres1)</code></pre>
<pre><code>## 
## out of 14578 with nonzero total read count
## adjusted p-value &lt; 0.05
## LFC &gt; 0 (up)       : 413, 2.8%
## LFC &lt; 0 (down)     : 696, 4.8%
## outliers [1]       : 0, 0%
## low counts [2]     : 2261, 16%
## (mean count &lt; 26)
## [1] see &#39;cooksCutoff&#39; argument of ?results
## [2] see &#39;independentFiltering&#39; argument of ?results</code></pre>
</div>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>head(lres1)</code></pre>
<pre><code>## log2 fold change (MAP): Time t2 vs t0 
## Wald test p-value: Time t2 vs t0 
## DataFrame with 6 rows and 6 columns
##                  baseMean log2FoldChange     lfcSE      stat    pvalue
##                 &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
## ENSG00000000003  490.0172      0.1663745 0.0903392  1.956524 0.0504034
## ENSG00000000419  817.7807      0.0494917 0.0846342  0.584068 0.5591746
## ENSG00000000457   82.0788      0.0925123 0.1143484  0.942577 0.3458972
## ENSG00000000460  356.0716     -0.0895216 0.0917478 -1.122002 0.2618616
## ENSG00000001036  919.6068      0.0257185 0.0746623  0.339198 0.7344609
## ENSG00000001084  529.5940      0.1890299 0.0795917  2.279655 0.0226281
##                      padj
##                 &lt;numeric&gt;
## ENSG00000000003  0.263505
## ENSG00000000419  0.830262
## ENSG00000000457  0.689946
## ENSG00000000460  0.612625
## ENSG00000001036  0.909639
## ENSG00000001084  0.159263</code></pre>
</div>
<p>We see that the number of genes up and down has not changed. But, we can see that the logFC distribution has changed in the density plot below.</p>
<pre class="r"><code>par(mfrow=c(1,2))
plot(density(na.omit(lres1$log2FoldChange)),main=&quot;log2FoldChange&quot;)
plot(density(na.omit(lres1$padj)),main=&quot;BH adj pval&quot;)
par(mfrow=c(1,1))</code></pre>
<p><img src="lab_dge_files/figure-html/unnamed-chunk-54-1.svg" width="768" style="display: block; margin: auto auto auto 0;" /></p>
<pre class="r"><code>plotCounts(d,gene=&quot;ENSG00000095370&quot;,intgroup=&quot;Time&quot;,normalized=T)</code></pre>
<p><img src="lab_dge_files/figure-html/unnamed-chunk-56-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
<p>This step does not change the total number of DE genes. This may be useful in downstream steps where genes need to be filtered down based on fold change or if fold change values are used in functional analyses such as GSEA.</p>
<p><i class="fas fa-comments"></i> How many DE genes do you get?</p>
</div>
</div>
</div>
<div id="visualisation" class="section level1">
<h1><span class="header-section-number">5</span> Visualisation</h1>
<p>In this section, we can explore some useful visualisation of the differential gene expression output.</p>
<div id="ma-plot" class="section level2">
<h2><span class="header-section-number">5.1</span> MA plot</h2>
<p>The MA plot shows mean expression vs log fold change for all genes. The <code>plotMA()</code> function from DESeq2 takes a results object as input. Differentially expressed genes are marked in red.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>DESeq2::plotMA(res1)</code></pre>
<img src="lab_dge_files/figure-html/unnamed-chunk-57-1.svg" width="480" style="display: block; margin: auto auto auto 0;" />
</div>
<p><i class="fas fa-comments"></i> How does this plot change if you set log fold change filtering to minimum value of 1. How does the plot change when you use <code>lfcShrink()</code>?</p>
</div>
<div id="volcano-plot" class="section level2">
<h2><span class="header-section-number">5.2</span> Volcano plot</h2>
<p>A volcano plot is similar to the MA plot. It plots log fold change vs adjusted p-values.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>ggplot()+
  geom_point(data=as.data.frame(res1),aes(x=log2FoldChange,y=-log10(padj)),col=&quot;grey80&quot;,alpha=0.5)+
  geom_point(data=filter(as.data.frame(res1),padj&lt;0.05),aes(x=log2FoldChange,y=-log10(padj)),col=&quot;red&quot;,alpha=0.7)+
  geom_hline(aes(yintercept=-log10(0.05)),alpha=0.5)+
  theme_bw()</code></pre>
<img src="lab_dge_files/figure-html/unnamed-chunk-58-1.svg" width="480" style="display: block; margin: auto auto auto 0;" />
</div>
<p>X axis denotes log fold change and the y axis denotes -log10 adjusted p-value. The adjusted p-value is transformed so that the smallest p-values appears at the top. The horizontal grey line denotes the significance threshold line. All genes above this line (coloured red as well) are considered significant.</p>
<p><i class="fas fa-comments"></i> Why is the y-axis (p-value) on a -log scale?</p>
</div>
<div id="counts-plot" class="section level2">
<h2><span class="header-section-number">5.3</span> Counts plot</h2>
<p>It can be a good idea to manually verify some of these genes by plotting out it’s actual read count values. We can use the function <code>plotCounts()</code> to visualise the data points for a gene of interest. Below, we see the counts before and after normalisation.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>plotCounts(d,gene=rownames(res1)[1],intgroup=&quot;Time&quot;,normalized=F)
plotCounts(d,gene=rownames(res1)[1],intgroup=&quot;Time&quot;,normalized=T)</code></pre>
<img src="lab_dge_files/figure-html/unnamed-chunk-59-1.svg" width="480" style="display: block; margin: auto auto auto 0;" /><img src="lab_dge_files/figure-html/unnamed-chunk-59-2.svg" width="480" style="display: block; margin: auto auto auto 0;" />
</div>
<p><i class="fas fa-comments"></i> By looking at the count plots, do you agree that the top DE genes differ significantly between the groups compared?</p>
</div>
</div>
<div id="session-info" class="section level1">
<h1><span class="header-section-number">6</span> Session info</h1>
<pre><code>## R version 4.0.2 (2020-06-22)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 18.04.5 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
## LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so
## 
## locale:
##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] pheatmap_1.0.12             edgeR_3.30.3               
##  [3] limma_3.44.3                DESeq2_1.28.1              
##  [5] SummarizedExperiment_1.18.2 DelayedArray_0.14.1        
##  [7] matrixStats_0.56.0          Biobase_2.48.0             
##  [9] GenomicRanges_1.40.0        GenomeInfoDb_1.24.2        
## [11] IRanges_2.22.2              S4Vectors_0.26.1           
## [13] BiocGenerics_0.34.0         biomaRt_2.44.1             
## [15] gridExtra_2.3               ggplot2_3.3.2              
## [17] formattable_0.2.0.1         kableExtra_1.2.1           
## [19] dplyr_1.0.2                 lubridate_1.7.9            
## [21] leaflet_2.0.3               yaml_2.2.1                 
## [23] fontawesome_0.1.0           captioner_2.2.3            
## [25] bookdown_0.20               knitr_1.29                 
## 
## loaded via a namespace (and not attached):
##  [1] nlme_3.1-148           bitops_1.0-6           bit64_4.0.5           
##  [4] RColorBrewer_1.1-2     webshot_0.5.2          progress_1.2.2        
##  [7] httr_1.4.2             tools_4.0.2            R6_2.4.1              
## [10] mgcv_1.8-31            DBI_1.1.0              colorspace_1.4-1      
## [13] withr_2.2.0            tidyselect_1.1.0       prettyunits_1.1.1     
## [16] bit_4.0.4              curl_4.3               compiler_4.0.2        
## [19] rvest_0.3.6            xml2_1.3.2             labeling_0.3          
## [22] scales_1.1.1           genefilter_1.70.0      askpass_1.1           
## [25] rappdirs_0.3.1         stringr_1.4.0          digest_0.6.25         
## [28] rmarkdown_2.3          XVector_0.28.0         pkgconfig_2.0.3       
## [31] htmltools_0.5.0        dbplyr_1.4.4           htmlwidgets_1.5.1     
## [34] rlang_0.4.7            rstudioapi_0.11        RSQLite_2.2.0         
## [37] farver_2.0.3           generics_0.0.2         jsonlite_1.7.0        
## [40] crosstalk_1.1.0.1      BiocParallel_1.22.0    RCurl_1.98-1.2        
## [43] magrittr_1.5           GenomeInfoDbData_1.2.3 Matrix_1.2-18         
## [46] Rcpp_1.0.5             munsell_0.5.0          lifecycle_0.2.0       
## [49] stringi_1.4.6          zlibbioc_1.34.0        BiocFileCache_1.12.1  
## [52] grid_4.0.2             blob_1.2.1             crayon_1.3.4          
## [55] lattice_0.20-41        splines_4.0.2          annotate_1.66.0       
## [58] hms_0.5.3              locfit_1.5-9.4         pillar_1.4.6          
## [61] geneplotter_1.66.0     XML_3.99-0.5           glue_1.4.2            
## [64] evaluate_0.14          vctrs_0.3.4            gtable_0.3.0          
## [67] openssl_1.4.2          purrr_0.3.4            tidyr_1.1.2           
## [70] assertthat_0.2.1       xfun_0.16              xtable_1.8-4          
## [73] survival_3.1-12        viridisLite_0.3.0      tibble_3.0.3          
## [76] AnnotationDbi_1.50.3   memoise_1.1.0          ellipsis_0.3.1</code></pre>
<hr />
</div>


<footer class="footer">
<div class="container footer-container">
<div class="row">
  
<div class="col-sm-8" style="text-align:left;vertical-align:middle;">
<p class="small" style="color:#bdbdbd;">
<b>2020</b> • NBIS • SciLifeLab
</p>
</div>

<div class="col-sm-4" style="text-align:right;vertical-align:middle;">
<p style="color:#bdbdbd;">
<span>
  <span class="footericon" style="padding-right:4px; padding-left:4px">
    <a href="https://nbis.se/">
      <svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 496 512"><path d="M248 8C111.03 8 0 119.03 0 256s111.03 248 248 248 248-111.03 248-248S384.97 8 248 8zm82.29 357.6c-3.9 3.88-7.99 7.95-11.31 11.28-2.99 3-5.1 6.7-6.17 10.71-1.51 5.66-2.73 11.38-4.77 16.87l-17.39 46.85c-13.76 3-28 4.69-42.65 4.69v-27.38c1.69-12.62-7.64-36.26-22.63-51.25-6-6-9.37-14.14-9.37-22.63v-32.01c0-11.64-6.27-22.34-16.46-27.97-14.37-7.95-34.81-19.06-48.81-26.11-11.48-5.78-22.1-13.14-31.65-21.75l-.8-.72a114.792 114.792 0 0 1-18.06-20.74c-9.38-13.77-24.66-36.42-34.59-51.14 20.47-45.5 57.36-82.04 103.2-101.89l24.01 12.01C203.48 89.74 216 82.01 216 70.11v-11.3c7.99-1.29 16.12-2.11 24.39-2.42l28.3 28.3c6.25 6.25 6.25 16.38 0 22.63L264 112l-10.34 10.34c-3.12 3.12-3.12 8.19 0 11.31l4.69 4.69c3.12 3.12 3.12 8.19 0 11.31l-8 8a8.008 8.008 0 0 1-5.66 2.34h-8.99c-2.08 0-4.08.81-5.58 2.27l-9.92 9.65a8.008 8.008 0 0 0-1.58 9.31l15.59 31.19c2.66 5.32-1.21 11.58-7.15 11.58h-5.64c-1.93 0-3.79-.7-5.24-1.96l-9.28-8.06a16.017 16.017 0 0 0-15.55-3.1l-31.17 10.39a11.95 11.95 0 0 0-8.17 11.34c0 4.53 2.56 8.66 6.61 10.69l11.08 5.54c9.41 4.71 19.79 7.16 30.31 7.16s22.59 27.29 32 32h66.75c8.49 0 16.62 3.37 22.63 9.37l13.69 13.69a30.503 30.503 0 0 1 8.93 21.57 46.536 46.536 0 0 1-13.72 32.98zM417 274.25c-5.79-1.45-10.84-5-14.15-9.97l-17.98-26.97a23.97 23.97 0 0 1 0-26.62l19.59-29.38c2.32-3.47 5.5-6.29 9.24-8.15l12.98-6.49C440.2 193.59 448 223.87 448 256c0 8.67-.74 17.16-1.82 25.54L417 274.25z"></path>
      </svg>
    </a>
    </span>
    <span class="footericon" style="padding-right:4px; padding-left:4px">
      <a href="https://twitter.com/NBISwe"><svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 512 512"><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"></path>
    </svg>
    </a>
    </span>
    <span class="footericon" style="padding-left:4px">
      <a href="https://www.linkedin.com/company/nbisweden/">
        <svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 448 512"><path d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3zM135.4 416H69V202.2h66.5V416zm-33.2-243c-21.3 0-38.5-17.3-38.5-38.5S80.9 96 102.2 96c21.2 0 38.5 17.3 38.5 38.5 0 21.3-17.2 38.5-38.5 38.5zm282.1 243h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9V416z"></path>
        </svg>
    </a>
    </span>
  </span>
</p>
</div>

</div>
</div>
</footer>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
