<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Kallisto &amp; Sleuth</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/anchor-sections-1.0/anchor-sections.css" rel="stylesheet" />
<script src="site_libs/anchor-sections-1.0/anchor-sections.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<link id="font-awesome-1-attachment" rel="attachment" href="site_libs/font-awesome-5.1.0/fonts/fontawesome-webfont.ttf"/>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="assets/lab.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
#TOC {
  display: none !important;
}
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img src="assets/logo.svg" id="logo" style="height:30px;margin:0;"/></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="home_schedule.html">Schedule</a>
</li>
<li>
  <a href="home_content.html">Content</a>
</li>
<li>
  <a href="home_precourse.html">Precourse</a>
</li>
<li>
  <a href="home_info.html">Info</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Kallisto &amp; Sleuth</h1>
<h3 class="subtitle">Workshop on RNA-Seq</h3>

</div>


<!-- rmd lab header -->
<p><br></p>
<p>This workflow runs transcript-level quantification using Kallisto followed by differential expression analysis using Sleuth.</p>
<pre class="r"><code># source download function
source(&quot;https://raw.githubusercontent.com/NBISweden/workshop-RNAseq/master/assets/scripts.R&quot;)</code></pre>
<div id="kallisto" class="section level1">
<h1><span class="header-section-number">1</span> Kallisto</h1>
<p><strong>Kallisto</strong> is an “alignment-free” RNA-Seq quantification method that runs very fast with a small memory footprint, so that it can be run on most laptops. It is a command-line program that can be downloaded as binary executables for Linux or Mac, or in source code format. For a first insight into the program, read <a href="https://liorpachter.wordpress.com/2015/05/10/near-optimal-rna-seq-quantification-with-kallisto/">here</a> and for the published article, see <a href="http://www.nature.com/nbt/journal/vaop/ncurrent/full/nbt.3519.html">here</a>. There is also a preprint <a href="http://arxiv.org/abs/1505.02710">here</a>.</p>
<p>Kallisto is geared towards quantification on the transcript (isoform) level, rather than the gene-level (although the latter can also be done by post-processing Kallisto output.) However, read assignment to transcript isoforms cannot (in general) be done unambiguously, so there is an intrinsic “quantification noise” or variability in this process. Kallisto can thus be run either in a single step (which is very fast) or in “bootstrap” mode (which takes longer, but can be done on several processors in parallel) in order to get uncertainty estimates for the expression levels - a kind of error bars for the quantification process. Running with bootstraps is mandatory if you want to perform differential expression analysis of isoforms with Sleuth (see below).</p>
<p>Kallisto is primarily meant for quantification of an existing set of FASTA sequences, that is, it does not perform transcript assembly and it cannot quantify the expression of novel transcripts that are not in the transcript index that you provide to it. With that said, you can of course use contigs from an assembly that you have produced in some other program in your Kallisto index. It would also be possible to use the software for eg: metagenomics or metatranscriptomics quantification.</p>
<div id="running-kallisto" class="section level2">
<h2><span class="header-section-number">1.1</span> Running Kallisto</h2>
<p>To load the Kallisto module on UPPMAX, execute</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>module load bioinfo-tools
module load kallisto/0.43.1</code></pre>
</div>
<p>Now we will download and merge human cDNA and ncDNA files from ENSEMBL in order to build a Kallisto transcript index. Note that we can concatenate the two gzipped files without unpacking them first. We use both the protein-coding transcripts and the non-coding ones to be able to capture more of the transcriptome.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>wget ftp://ftp.ensembl.org/pub/current_fasta/homo_sapiens/cdna/Homo_sapiens.GRCh38.cdna.all.fa.gz
wget ftp://ftp.ensembl.org/pub/current_fasta/homo_sapiens/ncrna/Homo_sapiens.GRCh38.ncrna.fa.gz
cat Homo_sapiens.GRCh38.cdna.all.fa.gz Homo_sapiens.GRCh38.ncrna.fa.gz &gt; Homo_sapiens.GRCh38.rna.fa.gz</code></pre>
</div>
<p>Now we can build the transcriptome index.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>kallisto index -i hsGRCh38_kallisto Homo_sapiens.GRCh38.rna.fa.gz</code></pre>
</div>
<p>It should take less than 10 minutes.</p>
<p>Next, copy the FASTA files from this uppmax path: <code>/proj/uppstore2017171/courses/RNAseqWorkshop/downloads/diffExp/FASTQ/</code>. When that is done, it’s time for quantifying the FASTQ files against our Kallisto index with bootstrapping (for later use in Sleuth). You could do that one by one, with a command like</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>kallisto quant -i hsGRCh38_kallisto \
  -t 4 \
  -b 100 \
  7_111116_AD0341ACXX_137_1_index1_1.fastq.gz 7_111116_AD0341ACXX_137_1_index1_2.fastq.gz \
  -o sample1</code></pre>
</div>
<p><strong>OR</strong> in a bash loop:</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>for i in {1..12};
  do
    kallisto quant -i hsGRCh38_kallisto -t 4 -b 100 7_111116_AD0341ACXX_137_${i}_index${i}_1.fastq.gz 7_111116_AD0341ACXX_137_${i}_index${i}_2.fastq.gz -o sample${i};
  done</code></pre>
</div>
<p>In this example, we put <code>-t 4</code> so we can use up to four processors in the bootstrapping. You may want to modify this value according to the machine you are working on. If you wanted to run Kallisto without bootstraps and just get expression values on a pair of FASTQ files, you would run something like</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>kallisto quant -i hsGRCh38_kallisto sample1_reads1.fastq sample1_reads2.fastq -o sample1</code></pre>
</div>
<p>Running Kallisto on all the 12 samples with 100 bootstraps may take an hour or so, depending on your machine and settings.</p>
</div>
</div>
<div id="sleuth" class="section level1">
<h1><span class="header-section-number">2</span> Sleuth</h1>
<p><strong>Sleuth</strong> is a companion package for Kallisto which is used for differential expression analysis of transcript quantifications from Kallisto. While you could use other differential expression packages such as <strong>Limma</strong> or <strong>DESeq2</strong> to analyse your Kallisto output, Sleuth also takes into consideration the inherent variability in transcript quantification as explained above. Sleuth also allows the modeling of covariates such as batch, individual, tissue type etc. in the same way as <strong>DESeq2/edgeR/Limma</strong>, which is useful for experimental designs with multiple varying factors.</p>
<p>Sleuth was designed to work on output from Kallisto (rather than count tables, like DESeq2, or BAM files, like CuffDiff2), so we need to run Kallisto first. (Note that the outputs from other RNA-seq quantifiers like <a href="https://github.com/COMBINE-lab/salmon">Salmon</a> or <a href="https://github.com/kingsfordgroup/sailfish">Sailfish</a> can also be used with Sleuth via the <a href="https://github.com/COMBINE-lab/wasabi">wasabi</a> package.)</p>
<p>Unlike Kallisto, Sleuth is an R package. At the end of a Sleuth analysis, it is possible to view a dynamical graphical presentation of the results where you can explore the differentially expressed transcripts in an intuitive way.</p>
<p>It is still early days for Sleuth and it has not been extensively benchmarked against other packages yet. Let’s try it on the same A431 data as in the DESeq2 lab!</p>
<div id="running-sleuth" class="section level2">
<h2><span class="header-section-number">2.1</span> Running Sleuth</h2>
<p>Here we give an example workflow for a DE analysis in Sleuth based on the A431 data that we are using for all the DE analysis labs. Start by copy the results from uppmax <code>/proj/uppstore2017171/courses/RNAseqWorkshop/downloads/diffExp/kallisto_results.tar.gz</code>. Download and extract the whole folder and make a note of where it is. We use the path <code>./data/kallisto_results/</code> relative to this document.</p>
<p>The Sleuth analysis is done entirely in R, so start your R environment and begin by installing the dependencies. This only needs to be done the first time, of course.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>source(&quot;http://bioconductor.org/biocLite.R&quot;)
biocLite(rhdf5)
install.packages(&quot;devtools&quot;)
library(devtools)
devtools::install_github(&quot;pachterlab/sleuth&quot;)</code></pre>
</div>
<p>We start by specifying paths to the Kallisto directories. We have the directory structure like below:</p>
<pre><code>kallisto_results\
  +-- sample1\
  |  +-- abundance.h5
  |  +-- abundance.tsv
  |  +-- run_info.json
  +-- sample2\
  +-- sample3\
...</code></pre>
<p>And we use the script below to generate the paths and labels.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>base_dir &lt;- &quot;./data/kallisto_results&quot;
samples &lt;- paste0(&quot;sample&quot;, 1:12)
kal_dirs &lt;- sapply(samples, function(id) file.path(base_dir, id))
kal_dirs</code></pre>
</div>
<pre><code>                           sample1                            sample2
 &quot;./data/kallisto_results/sample1&quot;  &quot;./data/kallisto_results/sample2&quot;
                           sample3                            sample4
 &quot;./data/kallisto_results/sample3&quot;  &quot;./data/kallisto_results/sample4&quot;
                           sample5                            sample6
 &quot;./data/kallisto_results/sample5&quot;  &quot;./data/kallisto_results/sample6&quot;
                           sample7                            sample8
 &quot;./data/kallisto_results/sample7&quot;  &quot;./data/kallisto_results/sample8&quot;
                           sample9                           sample10
 &quot;./data/kallisto_results/sample9&quot; &quot;./data/kallisto_results/sample10&quot;
                          sample11                           sample12
&quot;./data/kallisto_results/sample11&quot; &quot;./data/kallisto_results/sample12&quot;</code></pre>
<p>Now it’s time to fill in metadata about the samples. We can use a similar assignment as in the DESeq2 exercise.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>s2c &lt;- data.frame(path=kal_dirs,sample=samples,timepoint=rep(c(&quot;ctrl&quot;,&quot;t2h&quot;,&quot;t6h&quot;,&quot;t24h&quot;),each=3),stringsAsFactors=FALSE)</code></pre>
</div>
<p>Again, if there were other experimental factors involved, these could have been modelled here as well.</p>
<p>Now we read in the transcript-gene mappings file. This was <a href="lab_annotations.html">previously</a> downloaded.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>download_data(&quot;data/human_transcripts.txt&quot;)
t2g &lt;- read.delim(&quot;data/human_transcripts.txt&quot;, header=TRUE,sep=&quot;\t&quot;, stringsAsFactors=F)
t2g &lt;- dplyr::rename(t2g,target_id=ensembl_transcript_id,
                     ens_gene=ensembl_gene_id,ext_gene=external_gene_name)</code></pre>
</div>
<p>The next command will read the Kallisto output files, connect them with metadata, and set up a linear model for analyzing the expression data.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>library(&quot;sleuth&quot;)
so &lt;- sleuth_prep(s2c,~timepoint,target_mapping=t2g)</code></pre>
</div>
<pre><code>It appears that you are running Sleuth from within Rstudio.
Because of concerns with forking processes from a GUI, &#39;num_cores&#39; is being set to 1.
If you wish to take advantage of multiple cores, please consider running sleuth from the command line.reading in kallisto results
dropping unused factor levels
............
normalizing est_counts
47352 targets passed the filter
normalizing tpm
merging in metadata
summarizing bootstraps
............</code></pre>
<p>Next, we fit the linear model and test for one of the model coefficients. In this case we test the 24h time point versus the control.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>so &lt;- sleuth_fit(so)
so &lt;- sleuth_wt(so,which_beta=&quot;timepointt24h&quot;)</code></pre>
</div>
<pre><code>fitting measurement error models
shrinkage estimation
1 NA values were found during variance shrinkage estimation due to mean observation values outside of the range used for the LOESS fit.
The LOESS fit will be repeated using exact computation of the fitted surface to extrapolate the missing values.
These are the target ids with NA values: ENST00000465860
computing variance of betas</code></pre>
<p>Now we should be able to visualize the results:</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>sleuth_live(so)</code></pre>
</div>
<p>There are lots of things to look at here - explore according to your interests! Some things you might try are e.g. the PCA and sample heatmap options in the map menu, the test table in the analyses menu (which contains a ranked list of the differentially expressed genes), or the gene view in the same menu.</p>
<p>If you want to delve further into time series analysis with Sleuth (after all, we have just compared two time points here, whereas we have four in all), you might want to read this <a href="http://www.nxn.se/valent/timecourse-analysis-with-sleuth">excellent blog post</a> by Valentine Svensson. Note that Sleuth is still under development, so some of the commands may be a bit different.</p>
</div>
</div>
<div id="session-info" class="section level1">
<h1><span class="header-section-number">3</span> Session info</h1>
<pre><code>## R version 4.0.3 (2020-10-10)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 18.04.5 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
## LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so
## 
## locale:
##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] fgsea_1.16.0                enrichR_2.1                
##  [3] pvclust_2.2-0               rafalib_1.0.0              
##  [5] pheatmap_1.0.12             edgeR_3.32.0               
##  [7] limma_3.46.0                DESeq2_1.30.0              
##  [9] SummarizedExperiment_1.20.0 Biobase_2.50.0             
## [11] MatrixGenerics_1.2.0        matrixStats_0.57.0         
## [13] GenomicRanges_1.42.0        GenomeInfoDb_1.26.0        
## [15] IRanges_2.24.0              S4Vectors_0.28.0           
## [17] BiocGenerics_0.36.0         biomaRt_2.46.0             
## [19] gridExtra_2.3               ggplot2_3.3.2              
## [21] formattable_0.2.0.1         kableExtra_1.3.1           
## [23] dplyr_1.0.2                 lubridate_1.7.9            
## [25] leaflet_2.0.3               yaml_2.2.1                 
## [27] fontawesome_0.1.0           captioner_2.2.3            
## [29] bookdown_0.21               knitr_1.30                 
## 
## loaded via a namespace (and not attached):
##  [1] nlme_3.1-149           bitops_1.0-6           bit64_4.0.5           
##  [4] RColorBrewer_1.1-2     webshot_0.5.2          progress_1.2.2        
##  [7] httr_1.4.2             tools_4.0.3            R6_2.5.0              
## [10] mgcv_1.8-33            DBI_1.1.0              colorspace_2.0-0      
## [13] withr_2.3.0            tidyselect_1.1.0       prettyunits_1.1.1     
## [16] bit_4.0.4              curl_4.3               compiler_4.0.3        
## [19] rvest_0.3.6            xml2_1.3.2             DelayedArray_0.16.0   
## [22] labeling_0.4.2         scales_1.1.1           genefilter_1.72.0     
## [25] askpass_1.1            rappdirs_0.3.1         stringr_1.4.0         
## [28] digest_0.6.27          rmarkdown_2.5          XVector_0.30.0        
## [31] pkgconfig_2.0.3        htmltools_0.5.0        dbplyr_2.0.0          
## [34] htmlwidgets_1.5.2      rlang_0.4.8            rstudioapi_0.12       
## [37] RSQLite_2.2.1          farver_2.0.3           generics_0.1.0        
## [40] jsonlite_1.7.1         crosstalk_1.1.0.1      BiocParallel_1.24.1   
## [43] RCurl_1.98-1.2         magrittr_1.5           GenomeInfoDbData_1.2.4
## [46] Matrix_1.2-18          Rcpp_1.0.5             munsell_0.5.0         
## [49] lifecycle_0.2.0        stringi_1.5.3          zlibbioc_1.36.0       
## [52] BiocFileCache_1.14.0   grid_4.0.3             blob_1.2.1            
## [55] crayon_1.3.4           lattice_0.20-41        splines_4.0.3         
## [58] annotate_1.68.0        hms_0.5.3              locfit_1.5-9.4        
## [61] pillar_1.4.6           rjson_0.2.20           geneplotter_1.68.0    
## [64] fastmatch_1.1-0        XML_3.99-0.5           glue_1.4.2            
## [67] evaluate_0.14          data.table_1.13.2      vctrs_0.3.4           
## [70] gtable_0.3.0           openssl_1.4.3          purrr_0.3.4           
## [73] tidyr_1.1.2            assertthat_0.2.1       xfun_0.19             
## [76] xtable_1.8-4           survival_3.2-7         viridisLite_0.3.0     
## [79] tibble_3.0.4           AnnotationDbi_1.52.0   memoise_1.1.0         
## [82] ellipsis_0.3.1</code></pre>
<hr />
</div>


<footer class="footer">
<div class="container footer-container">
<div class="row">
  
<div class="col-sm-8" style="text-align:left;vertical-align:middle;">
<p class="small" style="color:#bdbdbd;">
<b>2020</b> • NBIS • SciLifeLab
</p>
</div>

<div class="col-sm-4" style="text-align:right;vertical-align:middle;">
<p style="color:#bdbdbd;">
<span>
  <span class="footericon" style="padding-right:4px; padding-left:4px">
    <a href="https://nbis.se/">
      <svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 496 512"><path d="M248 8C111.03 8 0 119.03 0 256s111.03 248 248 248 248-111.03 248-248S384.97 8 248 8zm82.29 357.6c-3.9 3.88-7.99 7.95-11.31 11.28-2.99 3-5.1 6.7-6.17 10.71-1.51 5.66-2.73 11.38-4.77 16.87l-17.39 46.85c-13.76 3-28 4.69-42.65 4.69v-27.38c1.69-12.62-7.64-36.26-22.63-51.25-6-6-9.37-14.14-9.37-22.63v-32.01c0-11.64-6.27-22.34-16.46-27.97-14.37-7.95-34.81-19.06-48.81-26.11-11.48-5.78-22.1-13.14-31.65-21.75l-.8-.72a114.792 114.792 0 0 1-18.06-20.74c-9.38-13.77-24.66-36.42-34.59-51.14 20.47-45.5 57.36-82.04 103.2-101.89l24.01 12.01C203.48 89.74 216 82.01 216 70.11v-11.3c7.99-1.29 16.12-2.11 24.39-2.42l28.3 28.3c6.25 6.25 6.25 16.38 0 22.63L264 112l-10.34 10.34c-3.12 3.12-3.12 8.19 0 11.31l4.69 4.69c3.12 3.12 3.12 8.19 0 11.31l-8 8a8.008 8.008 0 0 1-5.66 2.34h-8.99c-2.08 0-4.08.81-5.58 2.27l-9.92 9.65a8.008 8.008 0 0 0-1.58 9.31l15.59 31.19c2.66 5.32-1.21 11.58-7.15 11.58h-5.64c-1.93 0-3.79-.7-5.24-1.96l-9.28-8.06a16.017 16.017 0 0 0-15.55-3.1l-31.17 10.39a11.95 11.95 0 0 0-8.17 11.34c0 4.53 2.56 8.66 6.61 10.69l11.08 5.54c9.41 4.71 19.79 7.16 30.31 7.16s22.59 27.29 32 32h66.75c8.49 0 16.62 3.37 22.63 9.37l13.69 13.69a30.503 30.503 0 0 1 8.93 21.57 46.536 46.536 0 0 1-13.72 32.98zM417 274.25c-5.79-1.45-10.84-5-14.15-9.97l-17.98-26.97a23.97 23.97 0 0 1 0-26.62l19.59-29.38c2.32-3.47 5.5-6.29 9.24-8.15l12.98-6.49C440.2 193.59 448 223.87 448 256c0 8.67-.74 17.16-1.82 25.54L417 274.25z"></path>
      </svg>
    </a>
    </span>
    <span class="footericon" style="padding-right:4px; padding-left:4px">
      <a href="https://twitter.com/NBISwe"><svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 512 512"><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"></path>
    </svg>
    </a>
    </span>
    <span class="footericon" style="padding-left:4px">
      <a href="https://www.linkedin.com/company/nbisweden/">
        <svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 448 512"><path d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3zM135.4 416H69V202.2h66.5V416zm-33.2-243c-21.3 0-38.5-17.3-38.5-38.5S80.9 96 102.2 96c21.2 0 38.5 17.3 38.5 38.5 0 21.3-17.2 38.5-38.5 38.5zm282.1 243h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9V416z"></path>
        </svg>
    </a>
    </span>
  </span>
</p>
</div>

</div>
</div>
</footer>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
