<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Kallisto</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/anchor-sections-1.0/anchor-sections.css" rel="stylesheet" />
<script src="site_libs/anchor-sections-1.0/anchor-sections.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<link id="font-awesome-1-attachment" rel="attachment" href="site_libs/font-awesome-5.1.0/fonts/fontawesome-webfont.ttf"/>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="assets/lab.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
#TOC {
  display: none !important;
}
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img src="assets/logo.svg" id="logo" style="height:22px;margin:0;"/></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="home_schedule.html">Schedule</a>
</li>
<li>
  <a href="home_content.html">Content</a>
</li>
<li>
  <a href="home_precourse.html">Precourse</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Kallisto</h1>
<h3 class="subtitle">Workshop on RNA-Seq</h3>

</div>


<!-- rmd lab header -->
<p><br></p>
<div id="kallisto" class="section level1">
<h1><span class="header-section-number">1</span> Kallisto</h1>
<p><strong>Kallisto</strong> is an “alignment-free” RNA-Seq quantification method that runs very fast with a small memory footprint, so that it can be run on most laptops. It is a command-line program that can be downloaded as binary executables for Linux or Mac, or in source code format. For a first insight into the program, read <a href="https://liorpachter.wordpress.com/2015/05/10/near-optimal-rna-seq-quantification-with-kallisto/">here</a> and for the published article, see <a href="http://www.nature.com/nbt/journal/vaop/ncurrent/full/nbt.3519.html">here</a>.</p>
<p>Kallisto is geared towards quantification on the transcript (isoform) level, rather than the gene-level (although the latter can also be done by post-processing Kallisto output.) However, read assignment to transcript isoforms cannot (in general) be done unambiguously, so there is an intrinsic “quantification noise” or variability in this process. Kallisto can thus be run either in a single step (which is very fast) or in “bootstrap” mode (which takes longer, but can be done on several processors in parallel) in order to get uncertainty estimates for the expression levels - a kind of error bars for the quantification process. Running with bootstraps is mandatory if you want to perform differential expression analysis of isoforms with Sleuth (see below).</p>
<p>Kallisto is primarily meant for quantification of an existing set of FASTA sequences, that is, it does not perform transcript assembly and it cannot quantify the expression of novel transcripts that are not in the transcript index that you provide to it. With that said, you can of course use contigs from an assembly that you have produced in some other program in your Kallisto index. It would also be possible to use the software for eg: metagenomics or metatranscriptomics quantification.</p>
<p>Download the cDNA reference and create a index file to run on (This has already been done!)</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>#mkdir references/ref_kallisto
   
#cd references/ref_kallisto

#wget ftp://ftp.ensembl.org/pub/release-101/fasta/mus_musculus/cdna/Mus_musculus.GRCm38.cdna.all.fa.gz

#wait


#kallisto index \
#  --index kallisto_mm10_transcriptome_index.idx \
#  Mus_musculus.GRCm38.cdna.all.fa.gz

#wait

# cd ../..
</code></pre>
</div>
<p>It takes less than 10 minutes.</p>
<p>Now you can pseudoalign all your reads to the kalisto index using kalisto quantification</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>

  mkdir kallisto_quantification

 for sample_name in $(ls fastq/*.clean.fq.gz.1M_reads.fq.gz); do

    file=${sample_name#fastq/}
    sample=${file%.clean.fq.gz.1M_reads.fq.gz}
   
   echo $sample

     
    mkdir kallisto_quantification/$sample

    kallisto quant \
      --index references/ref_kallisto/kallisto_mm10_transcriptome_index.idx \
      --output-dir kallisto_quantification/$sample \
      --threads 3 \
      --single \
      --fragment-length 200 \
      --sd 20 \
     $sample_name

  done
</code></pre>
</div>
</div>
<div id="convert-from-transcript-estimation-to-gene-estimations." class="section level1">
<h1><span class="header-section-number">2</span> Convert from transcript estimation to gene estimations.</h1>
<p>This is done based on the <a href="https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html">txImport vignette</a></p>
<p>For the program to work it needs a trancript to gene converter. This is written in the cDNA names in the fasta file.</p>
<div class="block-title-parent">
<div class="block-title small">
sh
</div>
<pre class="sh"><code>   
gzcat references/ref_kallisto/Mus_musculus.GRCm38.cdna.all.fa.gz| \
grep &#39;&gt;&#39; |
awk &#39;{FS= &quot; &quot;}BEGIN{ print &quot;TXNAME,GENEID&quot;};{print substr($1,2) &quot;,&quot; substr($4,6)};&#39; \
&gt;references/ref_kallisto/tx2gene.mm.GRCm38.cdna.csv 
</code></pre>
</div>
<p>The next steps are carried out in R so make sure you start R with the correct packages.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>library(dplyr) # data wrangling
library(ggplot2) # plotting
library(DESeq2) # rna-seq
library(edgeR) # rna-seq
library(tximport) # importing kalisto transcript counts to geneLevels
library(readr) # Fast readr of files.
library(rhdf5) # read/convert kalisto output files.  


#if (!requireNamespace(&quot;BiocManager&quot;, quietly = TRUE))
#  install.packages(&quot;BiocManager&quot;)

#BiocManager::install(&quot;tximport&quot;)
#BiocManager::install(&quot;rhdf5&quot;)


# source download function
source(&quot;https://raw.githubusercontent.com/NBISweden/workshop-RNAseq/master/assets/scripts.R&quot;)

# download metadata

download_data(&quot;data/metadata_raw.csv&quot;)
mr &lt;- read.csv(&quot;data/metadata_raw.csv&quot;,header=TRUE,stringsAsFactors=F,row.names=1)</code></pre>
</div>
</div>
<div id="read-kalisto-files" class="section level1">
<h1><span class="header-section-number">3</span> Read Kalisto files</h1>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>files &lt;- paste(&quot;kallisto_quantification&quot; ,  
               list.files(path = &quot;kallisto_quantification&quot;,pattern = &quot;abundance.tsv&quot;, recursive = TRUE),
               sep = &quot;/&quot;)
names(files) &lt;- mr$SampleName


tx2gene &lt;- read_csv(&quot;references/ref_kallisto/tx2gene.mm.GRCm38.cdna.csv&quot;)


txi.kallisto.tsv &lt;- tximport(files, type = &quot;kallisto&quot;, tx2gene = tx2gene, ignoreAfterBar = TRUE)</code></pre>
</div>
</div>
<div id="convert-to-deseq2-object-for-analysis" class="section level1">
<h1><span class="header-section-number">4</span> Convert to DEseq2 object for analysis</h1>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>mr = mr %&gt;% mutate(Day = as.factor(Day))
dds &lt;- DESeqDataSetFromTximport(txi.kallisto.tsv, mr, ~Day)</code></pre>
</div>
</div>
<div id="convert-to-a-edger-object-for-analysis" class="section level1">
<h1><span class="header-section-number">5</span> Convert to a edgeR object for analysis</h1>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<pre class="r"><code>cts &lt;- txi.kallisto.tsv$counts
normMat &lt;- txi.kallisto.tsv$length
normMat &lt;- normMat/exp(rowMeans(log(normMat)))
library(edgeR)
o &lt;- log(calcNormFactors(cts/normMat)) + log(colSums(cts/normMat))
y &lt;- DGEList(cts)
y$offset &lt;- t(t(log(normMat)) + o)

# y is now ready for estimate dispersion functions see edgeR User&#39;s Guide</code></pre>
</div>
</div>
<div id="session-info" class="section level1">
<h1><span class="header-section-number">6</span> Session info</h1>
<pre><code>## R version 4.0.3 (2020-10-10)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 18.04.5 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
## LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so
## 
## locale:
##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] scales_1.1.1                fgsea_1.16.0               
##  [3] enrichR_2.1                 rafalib_1.0.0              
##  [5] pvclust_2.2-0               pheatmap_1.0.12            
##  [7] biomaRt_2.46.0              edgeR_3.32.0               
##  [9] limma_3.46.0                DESeq2_1.30.0              
## [11] SummarizedExperiment_1.20.0 Biobase_2.50.0             
## [13] MatrixGenerics_1.2.0        matrixStats_0.57.0         
## [15] GenomicRanges_1.42.0        GenomeInfoDb_1.26.0        
## [17] IRanges_2.24.0              S4Vectors_0.28.0           
## [19] BiocGenerics_0.36.0         ggplot2_3.3.2              
## [21] formattable_0.2.0.1         kableExtra_1.3.1           
## [23] dplyr_1.0.2                 lubridate_1.7.9.2          
## [25] leaflet_2.0.3               yaml_2.2.1                 
## [27] fontawesome_0.1.0           captioner_2.2.3            
## [29] bookdown_0.21               knitr_1.30                 
## 
## loaded via a namespace (and not attached):
##  [1] nlme_3.1-149           bitops_1.0-6           bit64_4.0.5           
##  [4] webshot_0.5.2          RColorBrewer_1.1-2     progress_1.2.2        
##  [7] httr_1.4.2             tools_4.0.3            R6_2.5.0              
## [10] DBI_1.1.0              mgcv_1.8-33            colorspace_2.0-0      
## [13] withr_2.3.0            gridExtra_2.3          prettyunits_1.1.1     
## [16] tidyselect_1.1.0       curl_4.3               bit_4.0.4             
## [19] compiler_4.0.3         rvest_0.3.6            xml2_1.3.2            
## [22] DelayedArray_0.16.0    labeling_0.4.2         genefilter_1.72.0     
## [25] askpass_1.1            rappdirs_0.3.1         stringr_1.4.0         
## [28] digest_0.6.27          rmarkdown_2.5          XVector_0.30.0        
## [31] pkgconfig_2.0.3        htmltools_0.5.0        dbplyr_2.0.0          
## [34] htmlwidgets_1.5.2      rlang_0.4.8            rstudioapi_0.13       
## [37] RSQLite_2.2.1          generics_0.1.0         farver_2.0.3          
## [40] jsonlite_1.7.1         crosstalk_1.1.0.1      BiocParallel_1.24.1   
## [43] RCurl_1.98-1.2         magrittr_1.5           GenomeInfoDbData_1.2.4
## [46] Matrix_1.2-18          Rcpp_1.0.5             munsell_0.5.0         
## [49] lifecycle_0.2.0        stringi_1.5.3          zlibbioc_1.36.0       
## [52] BiocFileCache_1.14.0   grid_4.0.3             blob_1.2.1            
## [55] crayon_1.3.4           lattice_0.20-41        splines_4.0.3         
## [58] annotate_1.68.0        hms_0.5.3              locfit_1.5-9.4        
## [61] pillar_1.4.6           rjson_0.2.20           geneplotter_1.68.0    
## [64] fastmatch_1.1-0        XML_3.99-0.5           glue_1.4.2            
## [67] evaluate_0.14          data.table_1.13.2      vctrs_0.3.4           
## [70] openssl_1.4.3          gtable_0.3.0           purrr_0.3.4           
## [73] tidyr_1.1.2            assertthat_0.2.1       xfun_0.19             
## [76] xtable_1.8-4           survival_3.2-7         viridisLite_0.3.0     
## [79] tibble_3.0.4           AnnotationDbi_1.52.0   memoise_1.1.0         
## [82] ellipsis_0.3.1</code></pre>
<hr />
</div>


<footer class="footer">
<div class="container footer-container">
<div class="row">
  
<div class="col-sm-8" style="text-align:left;vertical-align:middle;">
<p class="small" style="color:#bdbdbd;">
<b>2020</b> • NBIS • SciLifeLab
</p>
</div>

<div class="col-sm-4" style="text-align:right;vertical-align:middle;">
<p style="color:#bdbdbd;">
<span>
  <span class="footericon" style="padding-right:4px; padding-left:4px">
    <a href="https://nbis.se/">
      <svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 496 512"><path d="M248 8C111.03 8 0 119.03 0 256s111.03 248 248 248 248-111.03 248-248S384.97 8 248 8zm82.29 357.6c-3.9 3.88-7.99 7.95-11.31 11.28-2.99 3-5.1 6.7-6.17 10.71-1.51 5.66-2.73 11.38-4.77 16.87l-17.39 46.85c-13.76 3-28 4.69-42.65 4.69v-27.38c1.69-12.62-7.64-36.26-22.63-51.25-6-6-9.37-14.14-9.37-22.63v-32.01c0-11.64-6.27-22.34-16.46-27.97-14.37-7.95-34.81-19.06-48.81-26.11-11.48-5.78-22.1-13.14-31.65-21.75l-.8-.72a114.792 114.792 0 0 1-18.06-20.74c-9.38-13.77-24.66-36.42-34.59-51.14 20.47-45.5 57.36-82.04 103.2-101.89l24.01 12.01C203.48 89.74 216 82.01 216 70.11v-11.3c7.99-1.29 16.12-2.11 24.39-2.42l28.3 28.3c6.25 6.25 6.25 16.38 0 22.63L264 112l-10.34 10.34c-3.12 3.12-3.12 8.19 0 11.31l4.69 4.69c3.12 3.12 3.12 8.19 0 11.31l-8 8a8.008 8.008 0 0 1-5.66 2.34h-8.99c-2.08 0-4.08.81-5.58 2.27l-9.92 9.65a8.008 8.008 0 0 0-1.58 9.31l15.59 31.19c2.66 5.32-1.21 11.58-7.15 11.58h-5.64c-1.93 0-3.79-.7-5.24-1.96l-9.28-8.06a16.017 16.017 0 0 0-15.55-3.1l-31.17 10.39a11.95 11.95 0 0 0-8.17 11.34c0 4.53 2.56 8.66 6.61 10.69l11.08 5.54c9.41 4.71 19.79 7.16 30.31 7.16s22.59 27.29 32 32h66.75c8.49 0 16.62 3.37 22.63 9.37l13.69 13.69a30.503 30.503 0 0 1 8.93 21.57 46.536 46.536 0 0 1-13.72 32.98zM417 274.25c-5.79-1.45-10.84-5-14.15-9.97l-17.98-26.97a23.97 23.97 0 0 1 0-26.62l19.59-29.38c2.32-3.47 5.5-6.29 9.24-8.15l12.98-6.49C440.2 193.59 448 223.87 448 256c0 8.67-.74 17.16-1.82 25.54L417 274.25z"></path>
      </svg>
    </a>
    </span>
    <span class="footericon" style="padding-right:4px; padding-left:4px">
      <a href="https://twitter.com/NBISwe"><svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 512 512"><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"></path>
    </svg>
    </a>
    </span>
    <span class="footericon" style="padding-left:4px">
      <a href="https://www.linkedin.com/company/nbisweden/">
        <svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 448 512"><path d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3zM135.4 416H69V202.2h66.5V416zm-33.2-243c-21.3 0-38.5-17.3-38.5-38.5S80.9 96 102.2 96c21.2 0 38.5 17.3 38.5 38.5 0 21.3-17.2 38.5-38.5 38.5zm282.1 243h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9V416z"></path>
        </svg>
    </a>
    </span>
  </span>
</p>
</div>

</div>
</div>
</footer>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
